---
title: "Seurat Multiomics Workflow"
author: "PM"
date: "06/15/23"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
editor_options: 
  chunk_output_type: console
---

```{r, warning=FALSE, message=FALSE, warning=FALSE, echo=F}
graphics.off()
rm(list=ls())

#Set up R chunk options
knitr::opts_chunk$set(fig.width=8, fig.height=6,dpi=300,
                      echo=FALSE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(dev.args=list(bg="transparent"))

#Set directorires
ROOT_DIR<-"/fh/fast/furlan_s/experiments/timecourse_cd27"
stem<-"MO"
RMD_DIR  <- file.path(ROOT_DIR,  stem, "rmd")      # RMarkdown Directory
DATA_DIR <- file.path(ROOT_DIR,  stem, "data")      # Data Directory
RES_DIR  <- file.path(ROOT_DIR,  stem, "res")     # rds Directory
CDS_DIR <- file.path(ROOT_DIR,  stem, "cds")      #
FIG_DIR <- file.path(ROOT_DIR,  stem, "figs")     # figures directory
setwd("/fh/fast/furlan_s/experiments/timecourse_cd27/MO")

#Load all libraries
suppressPackageStartupMessages({
  library(monocle3)
  library(reticulate)
  library(openxlsx)  
  library(dplyr)
  library(Matrix)
  library(ggplot2)
  library(xfun)
  library(pals)
  library(RColorBrewer)
  library(Signac)
  library(Seurat)
  library(ggplot2)
  library(future)
  library(GenomicRanges)
  library(viridis)
  library(ComplexHeatmap)
  library(circlize)
  library(EnsDb.Hsapiens.v86)
  library(BSgenome.Hsapiens.UCSC.hg38)
  library(JASPAR2020)
  library(TFBSTools)
  library(patchwork)
  library(stringr)
  library(ggsignif)
  library(ggpubr)
  library(ggrastr)
  library(BuenColors)
  library(FigR)
  library(ggrepel)
  library(fgsea)
  library(ArchR)
  library(parallel)
  library(scCustomize)
  library(pbmcapply)
  library(FigR)
  library(xlsx)
})


set.seed(1234) # Set Seed

#h_cols <-rev(brewer.pal(name = "RdYlBu", n = 7)) # Color Palettes


#Some setup stuff, unsure exactly what a bitmap type is or array fire is
#options(bitmapType = 'cairo')


plan("multisession", workers = detectCores())
options(future.globals.maxSize = 700000 * 1024 ^ 3) # for 240 Gb RAM

dyn.load('/app/software/ArrayFire/3.8.1/lib64/libaf.so.3')
library(RcppArrayFire)
library(viewmastR)

py_config()
#cols<- c("#9893DA","#ebeb78","#fa9f42","#810e44","#d94745","#00B0E0","#3e5e8e","#0b6e4f","#b8e3ac")
#h_cols <-rev(brewer.pal(name = "RdYlBu", n = 7))


Sys.setenv(MODULEPATH="/app/modules/all", MODULEPATH_ROOT="/app/modules/all", MODULESHOME="/app/lmod/lmod")
source("/app/lmod/lmod/init/R", echo=FALSE, max=Inf)
module("load", "MACS2/2.2.6-foss-2019b-Python-3.7.4")
```

#Make peaks using macs2 for each sample(only run once)
```{bash}
ssh pmuthura@rhino
cd /fh/fast/furlan_s/grp/experiments/MO_cd27/outs
ml MACS2
mkdir peaks

macs2 callpeak -t atac_fragments.tsv.gz -g 2.7e+09 -f BED --nomodel --extsize 200 --shift -100 -n macs2 --outdir /fh/fast/furlan_s/grp/experiments/MO_cd27/outs/peaks
```



#Making each assay object and merge together
```{r Make objects and Merging data}
#First thing I need to do is arrange the data in the same organizational structure as Olivia's mm4 data
samps<-c("CJ_MO_P1", "CJ_MO_P2", "CJ_MO_P3", "CJ_MO_P4", "CJ_MO_P5")

#Loops through each sample, in output directory peaks (output from macs2)
macs_peaks<-lapply(samps, function(sample){
  peaks <- read.table(file = file.path(sample, "/outs/peaks/macs2_peaks.narrowPeak"))
  colnames(peaks)<-c("chr start end name integer_score none fold_enrichment log10pvalue log10qvalue relative_summit_pos") %>% strsplit(" ") %>% unlist()
  peaks
})

macs_granges<-lapply(macs_peaks, makeGRangesFromDataFrame)
macs_cgranges<-do.call(c, macs_granges)
#macs_mo.peaks <- reduce(x = macs_cgranges)
macs_mo.peaks <- GenomicRanges::reduce(macs_cgranges)


peakwidths <- width(macs_mo.peaks)
macs_mo.peaks <- macs_mo.peaks[peakwidths  < 10000 & peakwidths > 20]
macs_mo.peaks <- keepStandardChromosomes(macs_mo.peaks, pruning.mode = "coarse")

#Gets metadata for each sample
meta<-lapply(samps, function(sample){
  tab<-read.table(
  file = file.path(sample, "/outs/per_barcode_metrics.csv"),
  stringsAsFactors = FALSE,
  sep = ",",
  header = TRUE,
  row.names = 1
  )
  tab[tab$is_cell!=0, ]
})

#Builds atac matrix for each sample
frags<-lapply(1:length(samps), function(i){
  CreateFragmentObject(
    path = file.path(samps[i], "/outs/atac_fragments.tsv.gz"),
    cells = meta[[i]]$gex_barcode)
  })

#macs counts for each sample
macs_counts <- lapply(1:length(samps), function(i){
  FeatureMatrix(
  fragments = frags[[i]], process_n = 5000,
  features = macs_mo.peaks,
  cells = rownames(meta[[i]]))
})


counts<-macs_counts


#Package troubleshoot
#Sys.getenv("R_LIBS_USER") #Finds R env being used

# extract gene annotations from EnsDb
#library(EnsDb.Hsapiens.v86, lib.loc="/home/pmuthura/R/x86_64-pc-linux-gnu-library/4.2")
#library(GenomeInfoDb, lib.loc="/home/pmuthura/R/x86_64-pc-linux-gnu-library/4.2")
#library(GenomicFeatures, lib.loc="/home/pmuthura/R/x86_64-pc-linux-gnu-library/4.2")
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)


# change to UCSC style 
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "hg38"

#ATAC Assay
seus<-lapply(1:length(samps), function(i){
  ca<-CreateChromatinAssay(counts[[i]], fragments = frags[[i]])
  seu <- CreateSeuratObject(ca, assay = "ATAC", meta.data = meta[[i]])
  seu$dataset<-samps[i]
  seu
})

#troubleshooting
#i=3
#is.na(umi_counts) #Check for NAs
#duplicated(colnames(umi_counts)) %>% table #Check duplications in a table

#RNA and HTO Assays, HTO normalize and demux
seus<-lapply(1:length(samps), function(i){
  #RNA Assay
  raw<-Read10X_h5(file.path(samps[i], "/outs/filtered_feature_bc_matrix.h5")) #Reading in RNA
  seu<-CreateSeuratObject(
    counts = raw$`Gene Expression`[,colnames(counts[[i]])], assay= "RNA", meta.data = meta[[i]])
  seu[["ATAC"]]<-CreateChromatinAssay(counts[[i]], fragments = frags[[i]], sep = c(":", "-"), annotation = annotations)
  
  #HTO Assay
  umi_counts<-readMM(file.path(samps[i], "/outs/umi_count/matrix.mtx.gz")) #Reading in UMI counts
  colnames(umi_counts)<-data.table::fread(file.path(samps[i], "/outs/umi_count/barcodes.tsv.gz"), header = F)$V1
  rownames(umi_counts)<-data.table::fread(file.path(samps[i], "/outs/umi_count/features.tsv.gz"), header = F)$V1
  colnames(umi_counts)<-paste0(colnames(umi_counts), "-1")
  
  umi_subset<-umi_counts[1:8,which(colnames(umi_counts)%in%Cells(seu))] #Subsetting based on whether the counts are in the seurat object
  umi_final<-umi_subset[which(rowSums(umi_subset) > 3000),] #Removing hashes with trace counts (3000 or less)
  
  seu<-seu[,which(Cells(seu)%in%colnames(umi_final))] #Need to make sure the matrix dimensions match before adding new assay to seurat object
  
  #HTO Normalize and demux
  seu[["HTO"]]<-CreateAssayObject(umi_final)
  DefaultAssay(seu) <- "HTO"
  seu <- NormalizeData(seu, assay = "HTO", normalization.method = "CLR") #CLR: Applies a centered log ratio transformation
  seu <- HTODemux(seu, assay = "HTO", positive.quantile = 0.99)
  
  #seu[["HTO"]]<-CreateAssayObject(umi_counts[1:8,match(Cells(seu), colnames(umi_counts))])
  #seu[["HTO"]]<-CreateAssayObject(umi_subset)
  seu$dataset<-samps[i]
  seu
})


#table(Cells(MO) %in% paste0(colnames(counts), "-1")) ## didnt work, lot of cells in MO dont have HTO counts(didnt 


#Merge step
mo <- merge(
  x = seus[[1]],
  y = seus[2:length(seus)],
  add.cell.ids = samps
)

#Clearing data
#counts<-NULL
#macs_counts<-NULL
saveRDS(mo, file.path(CDS_DIR, "timecourse_MO_cd27_unfiltered.rds"))
mo<-readRDS(file.path(CDS_DIR, "timecourse_MO_cd27_unfiltered.rds"))

```

#HTO investigation(Just looking through some stats for Olivia)
```{r}
metadata <- mo_unfiltered@meta.data 

#creating new column identifier
metadata <- metadata %>% 
  rownames_to_column(var = "cell_id") %>% 
  mutate(identifier = stringr::str_extract(cell_id, "CJ_MO_P[1-5]"))

#nCount HTO per lane
result <- metadata %>% 
  group_by(dataset) %>% 
  summarise(total_nCount_HTO = sum(nCount_HTO, na.rm = TRUE))

print(result)

#Total num singlets
num_singlets <- metadata %>%
  dplyr::filter(HTO_classification.global == 'Singlet') %>%
  summarise(n = n())

print(num_singlets)

#num singlets per lane
num_singlets_per_identifier <- metadata %>%
  group_by(identifier) %>%
  dplyr::filter(HTO_classification.global == 'Singlet') %>%
  summarise(n = n())

print(num_singlets_per_identifier)

#Total # of identifiers
identifier_counts <- metadata %>%
  group_by(identifier) %>%
  summarise(n = n())

print(identifier_counts)

joined_data <- inner_join(identifier_counts, num_singlets_per_identifier, by = "identifier")

# And then calculate the percentage
joined_data <- joined_data %>%
  mutate(percentage = (n_singlet / total) * 100)

print(joined_data)


```

#QC Steps
```{r QC}
DefaultAssay(mo) <- "ATAC"

#Frip (Fragment Reads in Peaks)
mo$Frip<-mo$atac_peak_region_fragments/mo$atac_fragments
#Log scale the ATAC counts data
mo$log_ATAC<-log10(mo$nCount_ATAC)

# compute TSS enrichment score per cell
#File path ERROR, unable to add into the violin plots
mo <- TSSEnrichment(object = mo, fast = T)
mo <- NucleosomeSignal(mo)

DefaultAssay(mo) <- "RNA"
mo[["percent.mt"]] <- PercentageFeatureSet(mo, pattern = "^MT-")
mo$log_RNA<-log10(mo$nCount_RNA)

mo <- subset(
  x = mo,
  subset = 
    nucleosome_signal < 1 &
    TSS.enrichment < 9 &
    percent.mt <= 10 &
    log_ATAC >= 3 &
    log_RNA >= 3 &
    log_ATAC <= 4.5 &
    log_RNA <= 4.5 &
    Frip >= .4
)

VlnPlot(mo, features =  c("TSS.enrichment", "nucleosome_signal", "TSS.percentile", "Frip", "percent.mt", "log_ATAC", "log_RNA"), group.by = "timepoint", pt.size = 0)


#Removing cluster 18 for re-embedding (either works)

#no_18 <- subset(
#  x = mo_subset, idents = 18,
#  invert = T)

#no_18_2<-mo_subset[,mo_subset$seurat_clusters != 18]

```

#Creating and labeling metadata needed
```{r metadata labeling}
mo$id <- mo$orig.ident

#Naive
mo$id[mo$dataset == "CJ_MO_P1" & mo$HTO_maxID == 'HT5-AAGTATCGTTTCGCA']<- "Naive"

#24 hour
mo$id[mo$dataset == "CJ_MO_P1" & mo$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "5/0/5_24"
mo$id[mo$dataset == "CJ_MO_P1" & mo$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "5/0.2/5_24"
mo$id[mo$dataset == "CJ_MO_P1" & mo$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "5/0.2/0.8_24"
mo$id[mo$dataset == "CJ_MO_P1" & mo$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "5/0.2/0_24"
mo$id[mo$dataset == "CJ_MO_P2" & mo$HTO_maxID == 'HT8-CTCCTCTGCAATTAC']<- "BEAD_24"

#48 hour
mo$id[mo$dataset == "CJ_MO_P2" & mo$HTO_maxID == 'HT7-TGTCTTTCCTGCCAG']<- "5/0/5_48"
mo$id[mo$dataset == "CJ_MO_P2" & mo$HTO_maxID == 'HT6-GGTTGCCAGATGTCA']<- "5/0.2/5_48"
mo$id[mo$dataset == "CJ_MO_P2" & mo$HTO_maxID == 'HT5-AAGTATCGTTTCGCA']<- "5/0.2/0.8_48"
mo$id[mo$dataset == "CJ_MO_P3" & mo$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "5/0.2/0_48"
mo$id[mo$dataset == "CJ_MO_P3" & mo$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "BEAD_48"

#Day 4.5
mo$id[mo$dataset == "CJ_MO_P3" & mo$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "5/0/5_4.5"
mo$id[mo$dataset == "CJ_MO_P3" & mo$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "5/0.2/5_4.5"
mo$id[mo$dataset == "CJ_MO_P4" & mo$HTO_maxID == 'HT8-CTCCTCTGCAATTAC']<- "5/0.2/0.8_4.5"
mo$id[mo$dataset == "CJ_MO_P4" & mo$HTO_maxID == 'HT7-TGTCTTTCCTGCCAG']<- "5/0.2/0_4.5"
mo$id[mo$dataset == "CJ_MO_P4" & mo$HTO_maxID == 'HT6-GGTTGCCAGATGTCA']<- "BEAD_4.5"

#Day 9
mo$id[mo$dataset == "CJ_MO_P4" & mo$HTO_maxID == 'HT5-AAGTATCGTTTCGCA']<- "5/0/5_9"
mo$id[mo$dataset == "CJ_MO_P5" & mo$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "5/0.2/5_9"
mo$id[mo$dataset == "CJ_MO_P5" & mo$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "5/0.2/0.8_9"
mo$id[mo$dataset == "CJ_MO_P5" & mo$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "5/0.2/0_9"
mo$id[mo$dataset == "CJ_MO_P5" & mo$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "BEAD_9"


mo$timepoint <- "naive"

mo$timepoint[grep("24",mo$id)] <- "24hr"
mo$timepoint[grep("48",mo$id)] <- "48hr"
mo$timepoint[grep("4.5",mo$id)] <- "4.5d"
mo$timepoint[grep("9",mo$id)] <- "9d"

mo$timepoint <- factor(mo$timepoint,levels = c("Naive","24hr","48hr","4.5d","9d")) #Set levels (reorder for split dim plot)



#Change labeling of dosages(did this after running scrublet)

#24 hour
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P1" & mo_scrublet$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "3-27_24hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P1" & mo_scrublet$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "HD_24hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P1" & mo_scrublet$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "LD_24hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P1" & mo_scrublet$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "3-28_24hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P2" & mo_scrublet$HTO_maxID == 'HT8-CTCCTCTGCAATTAC']<- "BEAD_24hr"

#48 hour
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P2" & mo_scrublet$HTO_maxID == 'HT7-TGTCTTTCCTGCCAG']<- "3-27_48hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P2" & mo_scrublet$HTO_maxID == 'HT6-GGTTGCCAGATGTCA']<- "HD_48hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P2" & mo_scrublet$HTO_maxID == 'HT5-AAGTATCGTTTCGCA']<- "LD_48hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P3" & mo_scrublet$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "3-28_48hr"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P3" & mo_scrublet$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "BEAD_48hr"

#Day 4.5
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P3" & mo_scrublet$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "3-27_4.5d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P3" & mo_scrublet$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "HD_4.5d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P4" & mo_scrublet$HTO_maxID == 'HT8-CTCCTCTGCAATTAC']<- "LD_4.5d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P4" & mo_scrublet$HTO_maxID == 'HT7-TGTCTTTCCTGCCAG']<- "3-28_4.5d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P4" & mo_scrublet$HTO_maxID == 'HT6-GGTTGCCAGATGTCA']<- "BEAD_4.5d"

#Day 9
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P4" & mo_scrublet$HTO_maxID == 'HT5-AAGTATCGTTTCGCA']<- "3-27_9d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P5" & mo_scrublet$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "HD_9d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P5" & mo_scrublet$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "LD_9d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P5" & mo_scrublet$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "3-28_9d"
mo_scrublet$id[mo_scrublet$dataset == "CJ_MO_P5" & mo_scrublet$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "BEAD_9d"



saveRDS(mo_scrublet, file.path(CDS_DIR, "timecourse_MO_cd27_filtered_scrublet.rds"))

```

#ATAC UMAP
```{r reduce dimensions and run UMAP and plot}
DefaultAssay(mo) <- "ATAC"
mo <- RunTFIDF(mo)
mo <- FindTopFeatures(mo, min.cutoff = 'q0') #Find most frequently observed features
mo<- RunSVD(mo)
ElbowPlot(mo,reduction = 'lsi')
mo <- RunUMAP(mo, reduction = 'lsi', dims = 2:20, reduction.name = "ATAC_UMAP") #running dimensional reduction
DimPlot(mo,  group.by ="dataset", label=T, reduction = "ATAC_UMAP")+ggtitle("ATAC UMAP") #UMAP plot

#Feature plots for QC metrics (look by dataset and id)
FeaturePlot(mo, features = c("TSS.enrichment", "nucleosome_signal", "TSS.percentile", "Frip", "percent.mt", "log_ATAC", "log_RNA") , reduction = "ATAC_UMAP")

DimPlot(mo,  group.by ="id", label=T, reduction = "ATAC_UMAP",label.box = T,  cols = paletteDiscrete(values = levels(factor(mo$id))))+ggtitle("ATAC UMAP") #UMAP plot
DimPlot(mo,  group.by ="timepoint", label=T, reduction = "ATAC_UMAP",label.box = T,  cols = paletteDiscrete(values = levels(factor(mo$timepoint))))+ggtitle("ATAC UMAP") #UMAP plot

#Overlay both id and split by timepoint
#change cols to the color palette used in the experiment
DimPlot(mo,  group.by ="id", label=T, reduction = "ATAC_UMAP",split.by ="timepoint",label.box = T,  cols = paletteDiscrete(values = levels(factor(mo$id))))+ggtitle("ATAC UMAP") #UMAP plot


#Dim plot vs Feature Plot: Dimplot is for categorical -> dataset ; feature plot is for continuous -> logRNA & gene expression

```

#RNA UMAP
```{r}
DefaultAssay(mo) <- "RNA"
mo<-FindVariableFeatures(mo) #Identifies features that are outliers on a 'mean variability plot'.
mo<-ScaleData(mo) #Scale/Center data
mo <- SCTransform(mo, method = "glmGamPoi", vars.to.regress = "percent.mt") #A normalization method for single-cell UMI count data using a variance stabilizing transformation.
mo <- RunPCA(mo, features = VariableFeatures(mo))
ElbowPlot(mo)
ElbowPlot(mo, ndims = 30)
mo<-RunUMAP(mo, dims = 1:30, verbose = T, reduction.name = "SCT_UMAP") #UMAP for SCT method
DimPlot(mo,  group.by ="id", label=T, reduction = "SCT_UMAP")+ggtitle("SCT UMAP")

```


# Not running, Get this to work
#Removing doublets, cleaning up data
```{r scrublet}
library("SeuratDisk")
library("Seurat")
library("reticulate")
library("viewmastR")
use_python("~/.conda/envs/scenic/bin/python3")
py_config()

mo<-scrublet(no_18)
names(mo@meta.data)
FeaturePlot(mo, features = "doublet_scores", reduction = "wnn.umap", order = T)
DimPlot(mo, reduction = "wnn.umap")

#doublet scores greater than 0.3 seem high
#cluster 22 is low quality
VlnPlot(mo, features = "doublet_scores")

mo<-mo[,mo$doublet_scores < 0.2]
mo<-mo[,mo$seurat_clusters != 22]

mo <- FindMultiModalNeighbors(mo, reduction.list = list("pca", "lsi"), dims.list = list(1:39, 2:20), weighted.nn.name = "wnn.nn")
mo<- RunUMAP(mo, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
DimPlot(mo, reduction = "wnn.umap")
FeaturePlot(mo, features = "doublet_scores", reduction = "wnn.umap", order = T)
FeaturePlot(mo, features = "log_RNA", reduction = "wnn.umap", order = T)

mo <- FindClusters(mo, resolution = 0.4, graph.name = "wsnn", random.seed = 123)#
DimPlot(mo, reduction = "wnn.umap", order = T)

#saveRDS(mo, file.path(CDS_DIR, "timecourse_MO_cd27_filtered_scrublet.rds"))
mo_scrublet <-readRDS(file.path(CDS_DIR, "timecourse_MO_cd27_filtered_scrublet.rds"))
```

#WNN and Clustering Post Scrublet
```{r}
#WNN
mo_scrublet <- FindMultiModalNeighbors(mo_scrublet, reduction.list = list("pca", "lsi"), dims.list = list(1:39, 2:20), weighted.nn.name = "wnn.nn")
mo_scrublet <- RunUMAP(mo_scrublet, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
DimPlot(mo_scrublet, reduction = "wnn.umap")

DimPlot(mo_scrublet, group.by = "dataset", reduction = "wnn.umap", label = T, label.box = T)+ggtitle("WNN UMAP")
DimPlot(mo_scrublet, group.by = "id", reduction = "wnn.umap", label = T, label.box = T)+ggtitle("WNN UMAP")
DimPlot(mo_scrublet, group.by = "timepoint", reduction = "wnn.umap", label = T)+ggtitle("WNN UMAP")
DimPlot(mo_scrublet,  group.by ="id", label=T, reduction = "wnn.umap",split.by ="timepoint",label.size = 3,label.box = T)+ggtitle("WNN UMAP")

#Feature Plots for marker genes
FeaturePlot(mo_scrublet,features = c("TCF7","SELL","LEF1","IL2","IL2RA","FOXP1","MYC","STAT1","TBX21","EOMES","BACH1","FOS","JUN","MALAT1","CCR7","PRMT2","CD96","IFNG"), reduction = "wnn.umap")
FeaturePlot(mo_scrublet,features = c("IKZF1","IKZF2"), reduction = "wnn.umap")

#Clustering
mo_scrublet <- FindClusters(mo_scrublet, resolution = 0.4, graph.name = "wsnn", random.seed = 123)#
DimPlot(mo, reduction = "wnn.umap", order = T)
DimPlot(mo_scrublet, group.by = "seurat_clusters", reduction = "wnn.umap", label = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_subset$seurat_clusters)), set = "stallion2"))+ggtitle("Seurat clusters")&NoAxes()



#Color each cluster individually (in ID UMAP)
for (i in 1:length(levels(factor(mo_scrublet$id)))) {
  id_cols<-rep("gray80",length(levels(factor(mo_scrublet$id))))
  names(id_cols)<-levels(factor(mo_scrublet$id))
  id_cols[i]<-"red"
  pdf(paste0("/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/figs/",names(id_cols)[i],"_umap.pdf"),width=10,height=10)
  print(DimPlot(mo_scrublet, group.by = "id", reduction = "wnn.umap", label = F, cols = id_cols, label.box = F)+NoAxes()+ggtitle("WNN UMAP"))
  dev.off()
  }

#Find Markers
Idents(mo_scrublet)<-"seurat_clusters"
mo_scrublet.markers <- FindAllMarkers(mo_scrublet, only.pos = FALSE,  logfc.threshold = 0.25) #This step runs for a very long time on large datasets

top_n<- mo_scrublet.markers %>%
    group_by(cluster) %>%
    slice_max(n = 8, order_by = avg_log2FC)
```

#Clustering
```{r cluster}
#saveRDS(mo_subset, file.path(CDS_DIR, "timecourse_MO_cd27_filtered_subset.rds"))
mo_subset<-readRDS(file.path(CDS_DIR, "timecourse_MO_cd27_filtered_subset.rds"))
DefaultAssay(mo_subset)<-"SCT"
mo_subset <- FindClusters(mo_subset, resolution = 0.7, graph.name = "wsnn", random.seed = 123, algorithm = 3)##0.5 is the default resolution, if you increase the number, you will get more cluster

#QC on clusters
VlnPlot(mo_subset, features =  c("TSS.enrichment", "nucleosome_signal", "TSS.percentile", "Frip", "percent.mt", "log_ATAC", "log_RNA"), group.by = "seurat_clusters", pt.size = 0)

#Cluster plot
DimPlot(mo_subset, group.by = "seurat_clusters", reduction = "wnn.umap", label = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_subset$seurat_clusters)), set = "stallion2"))+ggtitle("Seurat clusters")&NoAxes()

#Feature plot for T-cell marker genes
FeaturePlot(mo_subset,features = c("CD4","CD8A","CD3E"), reduction = "wnn.umap")

#------------------------------------------------------------------------------------------------------------------------------------------------
#Clustering after scrublet(done in previous code block)



#------------------------------------------------------------------------------------------------------------------------------------------------

#MO$clusters<-as.character(MO$seurat_clusters)
#MO$clusters[MO$clusters %in% c(7)]<-"Naive"
#MO$clusters[MO$clusters %in% c(2)]<-"Naive"
#MO$clusters[MO$clusters %in% c(3,5)]<-"Naive"
#MO$clusters[MO$clusters %in% c(1,4)]<-"CD27_IKZF2-"
#MO$clusters[MO$clusters %in% c(8)]<-"CD27_IKZF2+"
#MO$clusters[MO$clusters %in% c(9)]<-"intermediate_cells"
#MO$clusters[MO$clusters %in% c(6)]<-"Bead_IL2+"
#MO$clusters[MO$clusters %in% c(0)]<-"Bead_IL2-"

#MO$clusters<-factor(MO$clusters, levels = c("Naive", "intermediate_cells", "CD27_IKZF2+", "CD27_IKZF2-", "Bead_IL2-", "Bead_IL2+"))
#MO$group<-factor(MO$group, levels = c("Naive", "24hr_3/27", "24hr_Beads"))
#DimPlot(MO, group.by = "clusters", reduction = "wnn.umap")

```

#Adding other Assays
```{r Gene Activity Assay}
#Ran this code block on terminal via grabnode as it was throwing error
mo_scrublet<-readRDS(file.path(CDS_DIR, "timecourse_MO_cd27_filtered_scrublet.rds"))

DefaultAssay(mo_scrublet)<- "ATAC"
gene.activities <- GeneActivity(mo_scrublet) #Compute counts per cell in gene body and promoter region.

mo_scrublet[['GENE_ACC']] <- CreateAssayObject(counts = gene.activities)

#Log normalize data, takes a very long time
mo_scrublet <- NormalizeData(
  object = mo_scrublet,
  assay = 'GENE_ACC',
  normalization.method = 'LogNormalize',
  scale.factor = median(mo_scrublet$nCount_GENE_ACC)
)
DefaultAssay(mo_scrublet)<- "GENE_ACC"
mo_scrublet<-ScaleData(mo_scrublet)

#Object doubled in size after these steps around 25 GB now
saveRDS(mo_scrublet, file.path(CDS_DIR, "timecourse_MO_cd27_filtered_scrublet.rds"))
```

```{r Add Motifs}
DefaultAssay(mo_scrublet)<- "ATAC"

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
mo_scrublet <- AddMotifs(
  object = mo_scrublet,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm
)
 

human_pwms_v3 <- readRDS("~/cisBP_human_pfms_2021.rds") #Can't find this file?
library(Signac)
DefaultAssay(mo_scrublet)<-"ATAC"

mo_scrublet@assays$ATAC@motifs<-NULL
mo_scrublet <- AddMotifs(
  object = mo_scrublet,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = human_pwms_v3
)
```

```{r Run ChromVAR}
#Runs for a long time, didn't work
mo_scrublet <- RunChromVAR(
  object = mo_scrublet,
  genome = BSgenome.Hsapiens.UCSC.hg38
)
rownames(mo_scrublet@assays$chromvar)

#Going to subset first
```

```{r save final RDS}
saveRDS(mo_scrublet, file.path(CDS_DIR, "timecourse_MO_cd27_master.rds"))
```

#=============================================START HERE=============================================
#Subset each timepoint 
```{r Subset each timepoint}
mo_scrublet<-readRDS(file.path(CDS_DIR, "timecourse_MO_cd27_filtered_scrublet.rds"))

mo_24_48 <- mo_scrublet[,mo_scrublet$id != 'LD_9d']
mo_24_48 <- mo_24_48[,mo_24_48$id != 'HD_9d']
mo_24_48 <- mo_24_48[,mo_24_48$id != 'BEAD_9d']
mo_24_48 <- mo_24_48[,mo_24_48$id != '3/28_9d']
mo_24_48 <- mo_24_48[,mo_24_48$id != '3/27_9d']

mo_24_48 <- mo_24_48[,mo_24_48$id != 'LD_4.5d']
mo_24_48 <- mo_24_48[,mo_24_48$id != 'HD_4.5d']
mo_24_48 <- mo_24_48[,mo_24_48$id != 'BEAD_4.5d']
mo_24_48 <- mo_24_48[,mo_24_48$id != '3/28_4.5d']
mo_24_48 <- mo_24_48[,mo_24_48$id != '3/27_4.5d']



mo_4.5_9 <- mo_scrublet[,mo_scrublet$id != 'LD_48hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != 'HD_48hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != 'BEAD_48hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != '3/28_48hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != '3/27_48hr']

mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != 'LD_24hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != 'HD_24hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != 'BEAD_24hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != '3/28_24hr']
mo_4.5_9 <- mo_4.5_9[,mo_4.5_9$id != '3/27_24hr']



mo_24hr <- mo_24_48[,mo_24_48$id != 'LD_48hr']
mo_24hr <- mo_24hr[,mo_24hr$id != 'HD_48hr']
mo_24hr <- mo_24hr[,mo_24hr$id != 'BEAD_48hr']
mo_24hr <- mo_24hr[,mo_24hr$id != '3/28_48hr']
mo_24hr <- mo_24hr[,mo_24hr$id != '3/27_48hr']

mo_48hr <- mo_24_48[,mo_24_48$id != 'LD_24hr']
mo_48hr <- mo_48hr[,mo_48hr$id != 'HD_24hr']
mo_48hr <- mo_48hr[,mo_48hr$id != 'BEAD_24hr']
mo_48hr <- mo_48hr[,mo_48hr$id != '3/28_24hr']
mo_48hr <- mo_48hr[,mo_48hr$id != '3/27_24hr']

mo_4.5d <- mo_4.5_9[,mo_4.5_9$id != 'LD_9d']
mo_4.5d <- mo_4.5d[,mo_4.5d$id != 'HD_9d']
mo_4.5d <- mo_4.5d[,mo_4.5d$id != 'BEAD_9d']
mo_4.5d <- mo_4.5d[,mo_4.5d$id != '3/28_9d']
mo_4.5d <- mo_4.5d[,mo_4.5d$id != '3/27_9d']

mo_9d <- mo_4.5_9[,mo_4.5_9$id != 'LD_4.5d']
mo_9d <- mo_9d[,mo_9d$id != 'HD_4.5d']
mo_9d <- mo_9d[,mo_9d$id != 'BEAD_4.5d']
mo_9d <- mo_9d[,mo_9d$id != '3/28_4.5d']
mo_9d <- mo_9d[,mo_9d$id != '3/27_4.5d']

mo_9d_no_naive <- mo_9d[,mo_9d$id != 'Naive']


```

# ATAC UMAP
```{r}
mo_24hr_og <- mo_24hr

#24hr
DefaultAssay(mo_24hr_og) <- "ATAC"
mo_24hr_og <- RunTFIDF(mo_24hr_og)
mo_24hr_og <- FindTopFeatures(mo_24hr_og, min.cutoff = 'q0') #Find most frequently observed features
mo_24hr_og<- RunSVD(mo_24hr_og)
ElbowPlot(mo_24hr_og,reduction = 'lsi')
mo_24hr_og <- RunUMAP(mo_24hr_og, reduction = 'lsi', dims = 2:20, reduction.name = "ATAC_UMAP") #running dimensional reduction
DimPlot(mo_24hr_og,  group.by ="id", label=T, reduction = "ATAC_UMAP")+ggtitle("ATAC UMAP") #UMAP plot

#48hr
DefaultAssay(mo_48hr) <- "ATAC"
mo_48hr <- RunTFIDF(mo_48hr)
mo_48hr <- FindTopFeatures(mo_48hr, min.cutoff = 'q0') #Find most frequently observed features
mo_48hr<- RunSVD(mo_48hr)
ElbowPlot(mo_48hr,reduction = 'lsi')
mo_48hr <- RunUMAP(mo_48hr, reduction = 'lsi', dims = 2:20, reduction.name = "ATAC_UMAP") #running dimensional reduction
DimPlot(mo_48hr,  group.by ="id", label=T, reduction = "ATAC_UMAP")+ggtitle("ATAC UMAP") #UMAP plot

#4.5d
DefaultAssay(mo_4.5d) <- "ATAC"
mo_4.5d <- RunTFIDF(mo_4.5d)
mo_4.5d <- FindTopFeatures(mo_4.5d, min.cutoff = 'q0') #Find most frequently observed features
mo_4.5d<- RunSVD(mo_4.5d)
ElbowPlot(mo_4.5d,reduction = 'lsi')
mo_4.5d <- RunUMAP(mo_4.5d, reduction = 'lsi', dims = 2:20, reduction.name = "ATAC_UMAP") #running dimensional reduction
DimPlot(mo_4.5d,  group.by ="id", label=T, reduction = "ATAC_UMAP")+ggtitle("ATAC UMAP") #UMAP plot

#9d
DefaultAssay(mo_9d) <- "ATAC"
mo_9d <- RunTFIDF(mo_9d)
mo_9d <- FindTopFeatures(mo_9d, min.cutoff = 'q0') #Find most frequently observed features
mo_9d<- RunSVD(mo_9d)
ElbowPlot(mo_9d,reduction = 'lsi')
mo_9d <- RunUMAP(mo_9d, reduction = 'lsi', dims = 2:20, reduction.name = "ATAC_UMAP") #running dimensional reduction
DimPlot(mo_9d,  group.by ="id", label=T, reduction = "ATAC_UMAP")+ggtitle("ATAC UMAP") #UMAP plot
```

# RNA UMAP
```{r}

#24hr
DefaultAssay(mo_24hr_og) <- "RNA"
mo_24hr_og<-FindVariableFeatures(mo_24hr_og) #Identifies features that are outliers on a 'mean variability plot'.
mo_24hr_og<-ScaleData(mo_24hr_og) #Scale/Center data
mo_24hr_og <- SCTransform(mo_24hr_og, method = "glmGamPoi", vars.to.regress = "percent.mt") #A normalization method for single-cell UMI count data using a variance stabilizing transformation.
mo_24hr_og <- RunPCA(mo_24hr_og, features = VariableFeatures(mo_24hr_og))
ElbowPlot(mo_24hr_og, ndims = 30)
mo_24hr_og<-RunUMAP(mo_24hr_og, dims = 1:30, verbose = T, reduction.name = "SCT_UMAP") #UMAP for SCT method
DimPlot(mo_24hr_og,  group.by ="id", label=T, reduction = "SCT_UMAP")+ggtitle("SCT UMAP")

#48hr
DefaultAssay(mo_48hr) <- "RNA"
mo_48hr<-FindVariableFeatures(mo_48hr) #Identifies features that are outliers on a 'mean variability plot'.
mo_48hr<-ScaleData(mo_48hr) #Scale/Center data
mo_48hr <- SCTransform(mo_48hr, method = "glmGamPoi", vars.to.regress = "percent.mt") #A normalization method for single-cell UMI count data using a variance stabilizing transformation.
mo_48hr <- RunPCA(mo_48hr, features = VariableFeatures(mo_48hr))
ElbowPlot(mo_48hr, ndims = 30)
mo_48hr<-RunUMAP(mo_48hr, dims = 1:30, verbose = T, reduction.name = "SCT_UMAP") #UMAP for SCT method
DimPlot(mo_48hr,  group.by ="id", label=T, reduction = "SCT_UMAP")+ggtitle("SCT UMAP")

#4.5d
DefaultAssay(mo_4.5d) <- "RNA"
mo_4.5d<-FindVariableFeatures(mo_4.5d) #Identifies features that are outliers on a 'mean variability plot'.
mo_4.5d<-ScaleData(mo_4.5d) #Scale/Center data
mo_4.5d <- SCTransform(mo_4.5d, method = "glmGamPoi", vars.to.regress = "percent.mt") #A normalization method for single-cell UMI count data using a variance stabilizing transformation.
mo_4.5d <- RunPCA(mo_4.5d, features = VariableFeatures(mo_4.5d))
ElbowPlot(mo_4.5d, ndims = 30)
mo_4.5d<-RunUMAP(mo_4.5d, dims = 1:30, verbose = T, reduction.name = "SCT_UMAP") #UMAP for SCT method
DimPlot(mo_4.5d,  group.by ="id", label=T, reduction = "SCT_UMAP")+ggtitle("SCT UMAP")

#9d
DefaultAssay(mo_9d) <- "RNA"
mo_9d<-FindVariableFeatures(mo_9d) #Identifies features that are outliers on a 'mean variability plot'.
mo_9d<-ScaleData(mo_9d) #Scale/Center data
mo_9d <- SCTransform(mo_9d, method = "glmGamPoi", vars.to.regress = "percent.mt") #A normalization method for single-cell UMI count data using a variance stabilizing transformation.
mo_9d <- RunPCA(mo_9d, features = VariableFeatures(mo_9d))
ElbowPlot(mo_9d, ndims = 30)
mo_9d<-RunUMAP(mo_9d, dims = 1:30, verbose = T, reduction.name = "SCT_UMAP") #UMAP for SCT method
DimPlot(mo_9d,  group.by ="id", label=T, reduction = "SCT_UMAP")+ggtitle("SCT UMAP")
```

#WNN UMAP
```{r}

#24hr
mo_24hr_og <- FindMultiModalNeighbors(mo_24hr_og, reduction.list = list("pca", "lsi"), dims.list = list(1:30, 2:10), weighted.nn.name = "wnn.nn")
mo_24hr_og <- RunUMAP(mo_24hr_og, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

DimPlot(mo_24hr_og, group.by = "id", reduction = "wnn.umap", label = T, label.box = T)+ggtitle("24hr WNN UMAP")

#48hr
mo_48hr <- FindMultiModalNeighbors(mo_48hr, reduction.list = list("pca", "lsi"), dims.list = list(1:30, 2:20), weighted.nn.name = "wnn.nn")
mo_48hr <- RunUMAP(mo_48hr, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

DimPlot(mo_48hr, group.by = "id", reduction = "wnn.umap", label = T, label.box = T)+ggtitle("48hr WNN UMAP")

#4.5d
mo_4.5d <- FindMultiModalNeighbors(mo_4.5d, reduction.list = list("pca", "lsi"), dims.list = list(1:30, 2:20), weighted.nn.name = "wnn.nn")
mo_4.5d <- RunUMAP(mo_4.5d, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

DimPlot(mo_4.5d, group.by = "id", reduction = "wnn.umap", label = T, label.box = T)+ggtitle("4.5d WNN UMAP")

#9d
mo_9d <- FindMultiModalNeighbors(mo_9d, reduction.list = list("pca", "lsi"), dims.list = list(1:30, 2:20), weighted.nn.name = "wnn.nn")
mo_9d <- RunUMAP(mo_9d, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

DimPlot(mo_9d, group.by = "id", reduction = "wnn.umap", label = T, label.box = T)+ggtitle("9d WNN UMAP")

```

# Seurat Clustering for each timepoint
```{r Seurat Clustering for each timepoint}

mo_24hr_og <- FindClusters(mo_24hr_og, resolution = 0.20, graph.name = "wsnn", random.seed = 123) #5 clusters

mo_48hr <- FindClusters(mo_48hr, resolution = 0.20, graph.name = "wsnn", random.seed = 123) #6 clusters

mo_4.5d <- FindClusters(mo_4.5d, resolution = 0.30, graph.name = "wsnn", random.seed = 123) #8 clusters

mo_9d <- FindClusters(mo_9d, resolution = 0.10, graph.name = "wsnn", random.seed = 123) #6 clusters

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 4 of 24hr object, then creating new metadata in the object based on this subset object

cluster4_24hr_data <- subset(mo_24hr_og, subset = seurat_clusters == 4)
cluster4_24hr_data <- FindNeighbors(cluster4_24hr_data, dims = 1:10, k.param = 5)
cluster4_24hr_data <- FindClusters(cluster4_24hr_data, resolution = 0.15, graph.name = "wsnn", random.seed = 123)

mo_24hr_og$sub_cluster_4 <- as.character(Idents(mo_24hr_og))
mo_24hr_og$sub_cluster_4[Cells(cluster4_24hr_data)] <- paste("c4",Idents(cluster4_24hr_data))


#checking sub-cluster 4 plot
DimPlot(mo_24hr_og, group.by = "sub_cluster_4", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_24hr_og$sub_cluster_4)), set = "stallion2"))+ggtitle("24hr Seurat clusters")&NoAxes()

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 0 of 24hr object, then creating new metadata in the object based on this subset object

cluster0_24hr_data <- subset(mo_24hr_og, subset = seurat_clusters == 0)
cluster0_24hr_data <- FindNeighbors(cluster0_24hr_data, dims = 1:10, k.param = 5)
cluster0_24hr_data <- FindClusters(cluster0_24hr_data, resolution = 0.15, graph.name = "wsnn", random.seed = 123)

mo_24hr_og$sub_cluster_0 <- as.character(Idents(mo_24hr_og))
mo_24hr_og$sub_cluster_0[Cells(cluster0_24hr_data)] <- paste("c4",Idents(cluster0_24hr_data))


#checking sub-cluster 4 plot
DimPlot(mo_24hr_og, group.by = "sub_cluster_0", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_24hr_og$sub_cluster_0)), set = "stallion2"))+ggtitle("24hr Seurat clusters")&NoAxes()

#------------------------------------------------------------------------------------------------------------

#Do this if we have more than 1 subcluster
#Add subcluster 0 embedding into subcluster 4 embedding
mo_24hr_og$sub_cluster_4[Cells(cluster0_24hr_data)] <- paste("c0",Idents(cluster0_24hr_data))

#Putting subclusters embedding back into normal embedding
mo_24hr_og$seurat_clusters <- mo_24hr_og$sub_cluster_4

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 2 of 24hr object, then creating new metadata in the object based on this subset object

cluster2_24hr_data <- subset(mo_24hr_og, subset = seurat_clusters == 2)
cluster2_24hr_data <- FindNeighbors(cluster2_24hr_data, dims = 1:10, k.param = 5)
cluster2_24hr_data <- FindClusters(cluster2_24hr_data, resolution = 0.15, graph.name = "wsnn", random.seed = 123)

mo_24hr_og$sub_cluster_2 <- as.character(Idents(mo_24hr_og))
mo_24hr_og$sub_cluster_2[Cells(cluster2_24hr_data)] <- paste("c2",Idents(cluster2_24hr_data))


#checking sub-cluster 2 plot
DimPlot(mo_24hr_og, group.by = "sub_cluster_2", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_24hr_og$sub_cluster_2)), set = "stallion2"))+ggtitle("24hr Seurat clusters")&NoAxes()

#Putting subclusters embedding back into normal embedding
mo_24hr_og$seurat_clusters <- mo_24hr_og$sub_cluster_2

#------------------------------------------------------------------------------------------------------------
#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 0 of 9d object, then creating new metadata in the object based on this subset object

cluster0_9d_data <- subset(mo_9d, subset = seurat_clusters == 0)
cluster0_9d_data <- FindNeighbors(cluster0_9d_data, dims = 1:10, k.param = 5)
cluster0_9d_data <- FindClusters(cluster0_9d_data, resolution = 0.15, graph.name = "wsnn", random.seed = 123)

mo_9d$sub_cluster_0 <- as.character(Idents(mo_9d))
mo_9d$sub_cluster_0[Cells(cluster0_9d_data)] <- paste("c0",Idents(cluster0_9d_data))


#checking sub-cluster 0 plot
DimPlot(mo_9d, group.by = "sub_cluster_0", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$sub_cluster_0)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 1 of 9d object, then creating new metadata in the object based on this subset object

cluster1_9d_data <- subset(mo_9d, subset = seurat_clusters == 1)
cluster1_9d_data <- FindNeighbors(cluster1_9d_data, dims = 1:10, k.param = 5)
cluster1_9d_data <- FindClusters(cluster1_9d_data, resolution = 0.10, graph.name = "wsnn", random.seed = 123)

mo_9d$sub_cluster_1 <- as.character(Idents(mo_9d))
mo_9d$sub_cluster_1[Cells(cluster1_9d_data)] <- paste("c1",Idents(cluster1_9d_data))

#checking sub-cluster 1 plot
DimPlot(mo_9d, group.by = "sub_cluster_1", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$sub_cluster_1)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()

#------------------------------------------------------------------------------------------------------------

#Add subcluster 1 embedding into subcluster 0 embedding
mo_9d$sub_cluster_0[Cells(cluster1_9d_data)] <- paste("c1",Idents(cluster1_9d_data))

#Putting subclusters embedding back into normal embedding
mo_9d$seurat_clusters <- mo_9d$sub_cluster_0

#------------------------------------------------------------------------------------------------------------

#Merging and renaming clusters
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c0 0"]<-"one"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c0 1"]<-"two"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c0 2"]<-"three"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c0 3"]<-"four"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "5"]<-"five"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "6"]<-"six"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c1 1"]<-"seven"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c1 0"]<-"eight"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "3"]<-"nine"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "2"]<-"ten"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "4"]<-"eleven"


mo_9d$seurat_clusters[mo_9d$seurat_clusters == "one"] <- "1"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "two"] <- "2"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "three"] <- "3"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "four"] <- "4"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "five"] <- "5"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "six"] <- "6"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "seven"] <- "7"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "eight"] <- "8"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "nine"] <- "9"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "ten"] <- "10"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "eleven"] <- "11"

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 2 of 9d object, then creating new metadata in the object based on this subset object

cluster2_9d_data <- subset(mo_9d, subset = seurat_clusters == 2)
cluster2_9d_data <- FindNeighbors(cluster2_9d_data, dims = 1:10, k.param = 5)
cluster2_9d_data <- FindClusters(cluster2_9d_data, resolution = 0.08, graph.name = "wsnn", random.seed = 123)

mo_9d$sub_cluster_2 <- as.character(Idents(mo_9d))
mo_9d$sub_cluster_2[Cells(cluster2_9d_data)] <- paste("c2",Idents(cluster2_9d_data))

#checking sub-cluster 2 plot
DimPlot(mo_9d, group.by = "sub_cluster_2", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$sub_cluster_2)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()

#Will put all new subclusters into sub_cluster_2 embedding and then back into normal embedding at the end.

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 11 of 9d object, then creating new metadata in the object based on this subset object

cluster11_9d_data <- subset(mo_9d, subset = seurat_clusters == 11)
cluster11_9d_data <- FindNeighbors(cluster11_9d_data, dims = 1:10, k.param = 5)
cluster11_9d_data <- FindClusters(cluster11_9d_data, resolution = 0.08, graph.name = "wsnn", random.seed = 123)

mo_9d$sub_cluster_11 <- as.character(Idents(mo_9d))
mo_9d$sub_cluster_11[Cells(cluster11_9d_data)] <- paste("c11",Idents(cluster11_9d_data))

#checking sub-cluster 11 plot
DimPlot(mo_9d, group.by = "sub_cluster_11", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$sub_cluster_11)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()

#Add subcluster 11 embedding into subcluster 2 embedding
mo_9d$sub_cluster_2[Cells(cluster11_9d_data)] <- paste("c11",Idents(cluster11_9d_data))

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 9 of 9d object, then creating new metadata in the object based on this subset object

cluster9_9d_data <- subset(mo_9d, subset = seurat_clusters == 9)
cluster9_9d_data <- FindNeighbors(cluster9_9d_data, dims = 1:10, k.param = 5)
cluster9_9d_data <- FindClusters(cluster9_9d_data, resolution = 0.20, graph.name = "wsnn", random.seed = 123)

mo_9d$sub_cluster_9 <- as.character(Idents(mo_9d))
mo_9d$sub_cluster_9[Cells(cluster9_9d_data)] <- paste("c9",Idents(cluster9_9d_data))

#checking sub-cluster 9 plot
DimPlot(mo_9d, group.by = "sub_cluster_9", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$sub_cluster_9)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()

#Add subcluster 9 embedding into subcluster 2 embedding
mo_9d$sub_cluster_2[Cells(cluster9_9d_data)] <- paste("c9",Idents(cluster9_9d_data))


#Putting subclusters embedding back into normal embedding
mo_9d$seurat_clusters <- mo_9d$sub_cluster_2

#------------------------------------------------------------------------------------------------------------

#Subclustering cluster 1 of 9d object, then creating new metadata in the object based on this subset object

cluster1_9d_data <- subset(mo_9d, subset = seurat_clusters == 1)
cluster1_9d_data <- FindNeighbors(cluster1_9d_data, dims = 1:10, k.param = 5)
cluster1_9d_data <- FindClusters(cluster1_9d_data, resolution = 0.20, graph.name = "wsnn", random.seed = 123)

mo_9d$sub_cluster_1 <- as.character(Idents(mo_9d))
mo_9d$sub_cluster_1[Cells(cluster1_9d_data)] <- paste("c1",Idents(cluster1_9d_data))

#checking sub-cluster 1 plot
DimPlot(mo_9d, group.by = "sub_cluster_1", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$sub_cluster_1)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()

#subclustering cluster 7 of 9d object
cluster7_9d_data <- subset(mo_9d, subset = seurat_clusters == 7)
cluster7_9d_data <- FindNeighbors(cluster7_9d_data, dims = 1:10, k.param = 5)
cluster7_9d_data <- FindClusters(cluster7_9d_data, resolution = 0.25, graph.name = "wsnn", random.seed = 123)

mo_9d$sub_cluster_7 <- as.character(Idents(mo_9d))
mo_9d$sub_cluster_1[Cells(cluster7_9d_data)] <- paste("c7",Idents(cluster7_9d_data))

#checking sub-cluster 1 plot
DimPlot(mo_9d, group.by = "sub_cluster_1", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$sub_cluster_1)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()

#Putting subclusters embedding back into normal embedding
mo_9d$seurat_clusters <- mo_9d$sub_cluster_1

#------------------------------------------------------------------------------------------------------------
#------------------------------------------------------------------------------------------------------------
#renaming and merging

mo_24hr_og$seurat_clusters[mo_24hr_og$seurat_clusters == "c2 0"] <- "2"
mo_24hr_og$seurat_clusters[mo_24hr_og$seurat_clusters == "c2 1"] <- "3"

mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c2 1"] <- "5"
mo_9d$seurat_clusters[mo_9d$seurat_clusters == "c2 0"] <- "2"

#------------------------------------------------------------------------------------------------------------

#Adding 24 hour plots for whole analysis 1/22/24
#Start here >>



saveRDS(mo_24hr, file.path(CDS_DIR, "24hr_naive_Jan24.rds"))
mo_24hr_og<-readRDS(file.path(CDS_DIR, "24hr_naive.rds"))


#6x8
#Checking cluster plots
DimPlot(mo_24hr_og, group.by = "seurat_clusters", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_24hr_og$seurat_clusters)), set = "stallion2"))+ggtitle("24hr Seurat clusters")&NoAxes()

DimPlot(mo_48hr, group.by = "seurat_clusters", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_48hr$seurat_clusters)), set = "stallion2"))+ggtitle("48hr Seurat clusters")&NoAxes()

DimPlot(mo_4.5d, group.by = "seurat_clusters", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_4.5d$seurat_clusters)), set = "stallion2"))+ggtitle("4.5d Seurat clusters")&NoAxes()

DimPlot(mo_9d, group.by = "seurat_clusters", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_9d$seurat_clusters)), set = "stallion2"))+ggtitle("9d Seurat clusters")&NoAxes()
```

#Renaming id's so they can be called in the red highlight loop below without breaking
```{r}
#24 hour
mo_24hr_og$id[mo_24hr_og$dataset == "CJ_MO_P1" & mo_24hr_og$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "3-27_24hr"
mo_24hr_og$id[mo_24hr_og$dataset == "CJ_MO_P1" & mo_24hr_og$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "HD_24hr"
mo_24hr_og$id[mo_24hr_og$dataset == "CJ_MO_P1" & mo_24hr_og$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "LD_24hr"
mo_24hr_og$id[mo_24hr_og$dataset == "CJ_MO_P1" & mo_24hr_og$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "3-28_24hr"
mo_24hr_og$id[mo_24hr_og$dataset == "CJ_MO_P2" & mo_24hr_og$HTO_maxID == 'HT8-CTCCTCTGCAATTAC']<- "BEAD_24hr"

#48 hour
mo_48hr$id[mo_48hr$dataset == "CJ_MO_P2" & mo_48hr$HTO_maxID == 'HT7-TGTCTTTCCTGCCAG']<- "3-27_48hr"
mo_48hr$id[mo_48hr$dataset == "CJ_MO_P2" & mo_48hr$HTO_maxID == 'HT6-GGTTGCCAGATGTCA']<- "HD_48hr"
mo_48hr$id[mo_48hr$dataset == "CJ_MO_P2" & mo_48hr$HTO_maxID == 'HT5-AAGTATCGTTTCGCA']<- "LD_48hr"
mo_48hr$id[mo_48hr$dataset == "CJ_MO_P3" & mo_48hr$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "3-28_48hr"
mo_48hr$id[mo_48hr$dataset == "CJ_MO_P3" & mo_48hr$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "BEAD_48hr"

#Day 4.5
mo_4.5d$id[mo_4.5d$dataset == "CJ_MO_P3" & mo_4.5d$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "3-27_4.5d"
mo_4.5d$id[mo_4.5d$dataset == "CJ_MO_P3" & mo_4.5d$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "HD_4.5d"
mo_4.5d$id[mo_4.5d$dataset == "CJ_MO_P4" & mo_4.5d$HTO_maxID == 'HT8-CTCCTCTGCAATTAC']<- "LD_4.5d"
mo_4.5d$id[mo_4.5d$dataset == "CJ_MO_P4" & mo_4.5d$HTO_maxID == 'HT7-TGTCTTTCCTGCCAG']<- "3-28_4.5d"
mo_4.5d$id[mo_4.5d$dataset == "CJ_MO_P4" & mo_4.5d$HTO_maxID == 'HT6-GGTTGCCAGATGTCA']<- "BEAD_4.5d"

#Day 9
mo_9d$id[mo_9d$dataset == "CJ_MO_P4" & mo_9d$HTO_maxID == 'HT5-AAGTATCGTTTCGCA']<- "3-27_9d"
mo_9d$id[mo_9d$dataset == "CJ_MO_P5" & mo_9d$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "HD_9d"
mo_9d$id[mo_9d$dataset == "CJ_MO_P5" & mo_9d$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "LD_9d"
mo_9d$id[mo_9d$dataset == "CJ_MO_P5" & mo_9d$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "3-28_9d"
mo_9d$id[mo_9d$dataset == "CJ_MO_P5" & mo_9d$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "BEAD_9d"
```

#Red highlight of each treatment group
```{r}
#Color each cluster individually (in ID UMAP)

#24hr
for (i in 1:length(levels(factor(mo_24hr_og$id)))) {
  id_cols<-rep("gray80",length(levels(factor(mo_24hr_og$id))))
  names(id_cols)<-levels(factor(mo_24hr_og$id))
  id_cols[i]<-"red"
  pdf(paste0("/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/figs/full_timecourse/Individual_Activation_Groups/",names(id_cols)[i],"_umap.pdf"),width=10,height=10)
  print(DimPlot(mo_24hr_og, group.by = "id", reduction = "wnn.umap", label = F, cols = id_cols, label.box = F)+NoAxes()+ggtitle("24 hr WNN UMAP"))
  dev.off()
}

#48hr
for (i in 1:length(levels(factor(mo_48hr$id)))) {
  id_cols<-rep("gray80",length(levels(factor(mo_48hr$id))))
  names(id_cols)<-levels(factor(mo_48hr$id))
  id_cols[i]<-"red"
  pdf(paste0("/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/figs/full_timecourse/",names(id_cols)[i],"_umap.pdf"),width=10,height=10)
  print(DimPlot(mo_48hr, group.by = "id", reduction = "wnn.umap", label = F, cols = id_cols, label.box = F)+NoAxes()+ggtitle("48 hr WNN UMAP"))
  dev.off()
}

#4.5d
for (i in 1:length(levels(factor(mo_4.5d$id)))) {
  id_cols<-rep("gray80",length(levels(factor(mo_4.5d$id))))
  names(id_cols)<-levels(factor(mo_4.5d$id))
  id_cols[i]<-"red"
  pdf(paste0("/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/figs/full_timecourse/",names(id_cols)[i],"_umap.pdf"),width=10,height=10)
  print(DimPlot(mo_4.5d, group.by = "id", reduction = "wnn.umap", label = F, cols = id_cols, label.box = F)+NoAxes()+ggtitle("4.5 d WNN UMAP"))
  dev.off()
}

#9d
for (i in 1:length(levels(factor(mo_9d$id)))) {
  id_cols<-rep("gray80",length(levels(factor(mo_9d$id))))
  names(id_cols)<-levels(factor(mo_9d$id))
  id_cols[i]<-"red"
  pdf(paste0("/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/figs/full_timecourse/",names(id_cols)[i],"_umap.pdf"),width=10,height=10)
  print(DimPlot(mo_9d, group.by = "id", reduction = "wnn.umap", label = F, cols = id_cols, label.box = F)+NoAxes()+ggtitle("9 d WNN UMAP"))
  dev.off()
  }

```

#Distribution Barplots
```{r}
#5x6
#Bar plot showing frequency of conditions within each cluster

#df$Var2 <- factor(df$Var2, levels = c("4","5","2","1","3","0")) #changing ordering for bar chart


#24hr
df_24hr <- data.frame(table(mo_24hr_og$id,mo_24hr_og$seurat_clusters))

ggplot(df_24hr, aes(fill = Var1, y=Freq, x=Var2)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Condition")) +
    scale_fill_manual(values = c("#186272","#f7ba97","red","#5b278c","#bd7ceb","gray"))+ggtitle("24hr Group Distrubution Barplot")

#changing ordering for bar chart
df_24hr$Var1 <- factor(df_24hr$Var1, levels = c("BEAD_24hr","LD_24hr","HD_24hr","3-27_24hr","3-28_24hr","Naive"))

ggplot(df_24hr, aes(fill = Var2, y=Freq, x=Var1)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Group") + 
    guides(fill=guide_legend(title="Clusters")) +
    scale_fill_manual(values = ArchR::paletteDiscrete(values = levels(factor(mo_24hr_og$seurat_clusters))))+ggtitle("24hr Cluster Distrubution Barplot")

#48hr
df_48hr <- data.frame(table(mo_48hr$id,mo_48hr$seurat_clusters))

ggplot(df_48hr, aes(fill = Var1, y=Freq, x=Var2)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Condition")) +
    scale_fill_manual(values = c("#186272","#f7ba97","red","#5b278c","#bd7ceb","gray"))+ggtitle("48hr Distrubution Barplot")

#changing ordering for bar chart
df_48hr$Var1 <- factor(df_48hr$Var1, levels = c("BEAD_48hr","LD_48hr","HD_48hr","3-27_48hr","3-28_48hr","Naive"))

ggplot(df_48hr, aes(fill = Var2, y=Freq, x=Var1)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Clusters")) +
    scale_fill_manual(values = ArchR::paletteDiscrete(values = levels(factor(mo_48hr$seurat_clusters))))+ggtitle("48hr Cluster Distrubution Barplot")




#4.5d
df_4.5d <- data.frame(table(mo_4.5d$id,mo_4.5d$seurat_clusters))

ggplot(df_4.5d, aes(fill = Var1, y=Freq, x=Var2)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Condition")) +
    scale_fill_manual(values = c("#186272","#f7ba97","red","#5b278c","#bd7ceb","gray"))+ggtitle("4.5d Distrubution Barplot")


#changing ordering for bar chart
df_4.5d$Var1 <- factor(df_4.5d$Var1, levels = c("BEAD_4.5d","LD_4.5d","HD_4.5d","3-27_4.5d","3-28_4.5d","Naive")) 

ggplot(df_4.5d, aes(fill = Var2, y=Freq, x=Var1)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Clusters")) +
    scale_fill_manual(values = ArchR::paletteDiscrete(values = levels(factor(mo_4.5d$seurat_clusters))))+ggtitle("4.5d Cluster Distrubution Barplot")

#9d
df_9d <- data.frame(table(mo_9d$id,mo_9d$seurat_clusters))

df_9d$Var2 <- factor(df_9d$Var2, levels = c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15"))

ggplot(df_9d, aes(fill = Var1, y=Freq, x=Var2)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Condition")) +
    scale_fill_manual(values = c("red","#bd7ceb","#5b278c","#186272","#f7ba97","gray"))+ggtitle("9d Distrubution Barplot")

#changing ordering for bar chart
df_9d$Var1 <- factor(df_9d$Var1, levels = c("BEAD_9d","LD_9d","HD_9d","3-27_9d","3-28_9d","Naive")) 

ggplot(df_9d, aes(fill = Var2, y=Freq, x=Var1)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Clusters")) +
    scale_fill_manual(values = ArchR::paletteDiscrete(values = levels(factor(mo_9d$seurat_clusters))))+ggtitle("9d Cluster Distrubution Barplot")


```

#Top DEGs(Differentially Expressed Genes) Dotplots
```{r}
#24hr
#Find Marker genes per cluster
DefaultAssay(mo_24hr_og) <- "RNA"
Idents(mo_24hr_og)<-"seurat_clusters"
mo_24hr.markers <- FindAllMarkers(mo_24hr_og, only.pos = FALSE,  logfc.threshold = 0.25) #This step runs for a very long time on large datasets


#changed from 12-->15
top_n_24hr<- mo_24hr.markers %>%
    group_by(cluster) %>%
    slice_max(n = 15, order_by = avg_log2FC)


library(svglite)
svglite::svglite("/fh/fast/furlan_s/experiments/timecourse_cd27/MO/figs/full_timecourse/cluster_dot_plot.svg", width = 15.5, height = 2)

#DotPlot(mo_naive_24hr, features = unique(top_n$gene), group.by = "test", cluster.idents=F)+scale_y_discrete(limits = c("1","2","3","4","5","6"))+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))&RotatedAxis()

#3x13
DotPlot(mo_24hr_og, features = unique(top_n_24hr$gene), group.by = "seurat_clusters", cluster.idents=F)+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))+theme(axis.text.x = element_text(color = "grey20", size = 8, angle = 90))&RotatedAxis()
dev.off()


#48hr
#Find Marker genes per cluster
Idents(mo_48hr)<-"seurat_clusters"
mo_48hr.markers <- FindAllMarkers(mo_48hr, only.pos = FALSE,  logfc.threshold = 0.25) #This step runs for a very long time on large datasets

#changed from 8-->12
top_n_48hr<- mo_48hr.markers %>%
    group_by(cluster) %>%
    slice_max(n = 12, order_by = avg_log2FC)

library(svglite)
svglite::svglite("/fh/fast/furlan_s/experiments/timecourse_cd27/MO/figs/full_timecourse/cluster_dot_plot.svg", width = 15.5, height = 2)

#DotPlot(mo_naive_24hr, features = unique(top_n$gene), group.by = "test", cluster.idents=F)+scale_y_discrete(limits = c("1","2","3","4","5","6"))+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))&RotatedAxis()

#3x13
DotPlot(mo_48hr, features = unique(top_n_48hr$gene), group.by = "seurat_clusters", cluster.idents=F)+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))+theme(axis.text.x = element_text(color = "grey20", size = 8, angle = 90))&RotatedAxis()
dev.off()


#4.5d
#Find Marker genes per cluster
Idents(mo_4.5d)<-"seurat_clusters"
mo_4.5d.markers <- FindAllMarkers(mo_4.5d, only.pos = FALSE,  logfc.threshold = 0.25) #This step runs for a very long time on large datasets

#changed from 8-->12
top_n_4.5d<- mo_4.5d.markers %>%
    group_by(cluster) %>%
    slice_max(n = 12, order_by = avg_log2FC)

library(svglite)
svglite::svglite("/fh/fast/furlan_s/experiments/timecourse_cd27/MO/figs/full_timecourse/cluster_dot_plot.svg", width = 15.5, height = 2)

#DotPlot(mo_naive_24hr, features = unique(top_n$gene), group.by = "test", cluster.idents=F)+scale_y_discrete(limits = c("1","2","3","4","5","6"))+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))&RotatedAxis()

#4x13
DotPlot(mo_4.5d, features = unique(top_n_4.5d$gene), group.by = "seurat_clusters", cluster.idents=F)+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))+theme(axis.text.x = element_text(color = "grey20", size = 8, angle = 90))&RotatedAxis()
dev.off()


#9d
#Find Marker genes per cluster
Idents(mo_9d)<-"seurat_clusters"
mo_9d.markers <- FindAllMarkers(mo_9d, only.pos = FALSE,  logfc.threshold = 0.25) #This step runs for a very long time on large datasets

#changed from 8-->12
top_n_9d<- mo_9d.markers %>%
    group_by(cluster) %>%
    slice_max(n = 12, order_by = avg_log2FC)

library(svglite)
svglite::svglite("/fh/fast/furlan_s/experiments/timecourse_cd27/MO/figs/full_timecourse/cluster_dot_plot.svg", width = 15.5, height = 2)

#DotPlot(mo_naive_24hr, features = unique(top_n$gene), group.by = "test", cluster.idents=F)+scale_y_discrete(limits = c("1","2","3","4","5","6"))+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))&RotatedAxis()

#5x23
DotPlot(mo_9d, features = unique(top_n_9d$gene), group.by = "seurat_clusters", cluster.idents=F)+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))+theme(axis.text.x = element_text(color = "grey20", size = 8, angle = 90))&RotatedAxis()

#group by id
DotPlot(mo_9d, features = unique(top_n_9d$gene), group.by = "id", cluster.idents=F)+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))&RotatedAxis()
dev.off()

```

#Saving and Loading Objects
```{r}
saveRDS(mo_24hr, file.path(CDS_DIR, "24hr_naive.rds"))
saveRDS(mo_48hr, file.path(CDS_DIR, "48hr_naive.rds"))
saveRDS(mo_4.5d, file.path(CDS_DIR, "4.5d_naive.rds"))
saveRDS(mo_9d, file.path(CDS_DIR, "9d_naive.rds"))

mo_24hr<-readRDS(file.path(CDS_DIR, "24hr_naive.rds"))
mo_48hr<-readRDS(file.path(CDS_DIR, "48hr_naive.rds"))
mo_4.5d<-readRDS(file.path(CDS_DIR, "4.5d_naive.rds"))
mo_9d<-readRDS(file.path(CDS_DIR, "9d_naive.rds"))
```

```{r}

```



#Motifs from JASPAR/cisBP and Chromvar per timepoint
```{r}
##JASPAR###

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE))

#48hr
DefaultAssay(mo_48hr)<- "ATAC"

# add motif information from JASPAR
mo_48hr@assays$ATAC@motifs<-NULL
mo_48hr <- AddMotifs(
  object = mo_48hr,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm
)
#rownames(mo_48hr[["chromvar"]])

mo_48hr <- RunChromVAR(
  object = mo_48hr,
  genome = BSgenome.Hsapiens.UCSC.hg38
)

#4.5d
DefaultAssay(mo_4.5d)<- "ATAC"

mo_4.5d@assays$ATAC@motifs<-NULL

# add motif information from JASPAR
mo_4.5d <- AddMotifs(
  object = mo_4.5d,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm
)
#rownames(mo_4.5d[["chromvar"]])

mo_4.5d <- RunChromVAR(
  object = mo_4.5d,
  genome = BSgenome.Hsapiens.UCSC.hg38
)

#9d
DefaultAssay(mo_9d)<- "ATAC"

mo_9d@assays$ATAC@motifs<-NULL

# add motif information from JASPAR
mo_9d <- AddMotifs(
  object = mo_9d,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm
)
#rownames(mo_9d[["chromvar"]])

mo_9d <- RunChromVAR(
  object = mo_9d,
  genome = BSgenome.Hsapiens.UCSC.hg38
)


library(Signac)
##cisBP##
human_pwms_v3 <- readRDS("~/cisBP_human_pfms_2021.rds")


#48hr
mo_48hr@assays$ATAC@motifs<-NULL
DefaultAssay(mo_48hr)<- "ATAC"

mo_48hr <- AddMotifs(
  object = mo_48hr,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = human_pwms_v3
)

mo_48hr <- RunChromVAR(
  object = mo_48hr,
  genome = BSgenome.Hsapiens.UCSC.hg38
)

#4.5d
mo_4.5d@assays$ATAC@motifs<-NULL
DefaultAssay(mo_4.5d)<- "ATAC"

mo_4.5d <- AddMotifs(
  object = mo_4.5d,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = human_pwms_v3
)

mo_4.5d <- RunChromVAR(
  object = mo_4.5d,
  genome = BSgenome.Hsapiens.UCSC.hg38
)

#9d
mo_9d@assays$ATAC@motifs<-NULL
DefaultAssay(mo_9d)<- "ATAC"

mo_9d <- AddMotifs(
  object = mo_9d,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = human_pwms_v3
)

mo_9d <- RunChromVAR(
  object = mo_9d,
  genome = BSgenome.Hsapiens.UCSC.hg38
)


#rownames(mo_9d@assays$chromvar)
```

#9d TF Dot Plots using CisBP
```{r}

#9d
DefaultAssay(mo_9d)<- "ATAC"
mo_9d@assays$ATAC@motifs<-NULL
mo_9d <- AddMotifs(
  object = mo_9d,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = human_pwms_v3 #cisBP 
)

mo_9d <- RunChromVAR(
  object = mo_9d,
  genome = BSgenome.Hsapiens.UCSC.hg38
)
rownames(MO@assays$chromvar)


DefaultAssay(mo_9d)<- "chromvar"
Idents(mo_9d)<-"id"
differential.activity.id <- FindAllMarkers(
  object = mo_9d,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)


t<- c()
for(i in 1:length(differential.activity.id$gene)){
  t[i]<-mo_9d@assays$ATAC@motifs@motif.names[[differential.activity.id$gene[i]]]
}
differential.activity.id$TF <- t

#select significant TFS
top_tf<- differential.activity.id %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
all_values.id<- differential.activity.id[which(differential.activity.id$TF %in% c(top_tf$TF)),]

all_values.id$cluster<-factor(all_values.id$cluster, levels = c("BEAD_9d", "LD_9d", "HD_9d","3-27_9d","3-28_9d"))
all_values.id<-all_values.id %>% mutate(cluster=factor(cluster, levels = c("BEAD_9d", "LD_9d", "HD_9d","3-27_9d","3-28_9d"))) 
all_values.id<-all_values.id %>% arrange(cluster, desc(avg_diff))
all_values.id$TF<-factor(all_values.id$TF, levels =rev(unique(all_values.id$TF)))
#Removes NA
all_values.id <- na.omit(all_values.id)

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
all_values.id$TF<-factor(all_values.id$TF, levels = levels(all_values.id$TF)[c(1:15,16:47,48:62,63:71,72,73:88)])

ggplot(data = all_values.id, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()


#per cluster
Idents(mo_9d)<-"seurat_clusters"
differential.activity.cluster <- FindAllMarkers(
  object = mo_9d,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

t<- c()
for(i in 1:length(differential.activity.cluster$gene)){
  t[i]<-mo_9d@assays$ATAC@motifs@motif.names[[differential.activity.cluster$gene[i]]]
}
differential.activity.cluster$TF <- t

#select significant TFS
top_tf_clusters<- differential.activity.cluster %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
all_values.clusters<- differential.activity.cluster[which(differential.activity.cluster$TF %in% c(top_tf_clusters$TF)),]

all_values.clusters$cluster<-factor(all_values.clusters$cluster, levels = c("1", "2", "3","4","15"))
all_values.clusters<-all_values.clusters %>% mutate(cluster=factor(cluster, levels = c("1", "2", "3","4","15"))) 
all_values.clusters<-all_values.clusters %>% arrange(cluster, desc(avg_diff))
all_values.clusters$TF<-factor(all_values.clusters$TF, levels =rev(unique(all_values.clusters$TF)))
#Removes NA
all_values.clusters <- na.omit(all_values.clusters)


##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
all_values.clusters$TF<-factor(all_values.clusters$TF, levels = levels(all_values.clusters$TF)[c(1:14,15:26,27:40,41:46,47:59,60:72,73:111,112:165,166:207,208:252:253:287,288:315,316:360,361:395,396:410)])

ggplot(data = all_values.clusters, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()


```

#9d TF Dot Plots using JASPAR
```{r}
DefaultAssay(mo_9d)<- "chromvar"
Idents(mo_9d)<-"id"

JASPAR.DA.id_9d <- FindAllMarkers(
  object = mo_9d,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

t<- c()
for(i in 1:length(JASPAR.DA.id_9d$gene)){
  t[i]<-mo_9d@assays$ATAC@motifs@motif.names[[JASPAR.DA.id_9d$gene[i]]]
}


JASPAR.DA.id_9d$TF <- t

#select significant TFS
top_tf_ids_9d<- JASPAR.DA.id_9d %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
JASPAR.all_values.id_9d<- JASPAR.DA.id_9d[which(JASPAR.DA.id_9d$gene %in% c(top_tf_ids_9d$gene)),]

JASPAR.all_values.id_9d$cluster<-factor(JASPAR.all_values.id_9d$cluster, levels = c("BEAD_9d", "LD_9d", "HD_9d","3-27_9d","3-28_9d"))
JASPAR.all_values.id_9d<-JASPAR.all_values.id_9d %>% mutate(cluster=factor(cluster, levels = c("BEAD_9d", "LD_9d", "HD_9d","3-27_9d","3-28_9d"))) 
JASPAR.all_values.id_9d<-JASPAR.all_values.id_9d %>% arrange(cluster, desc(avg_diff))
JASPAR.all_values.id_9d$gene<-factor(JASPAR.all_values.id_9d$gene, levels =rev(unique(JASPAR.all_values.id_9d$gene)))
#Removes NA
JASPAR.all_values.id_9d <- na.omit(JASPAR.all_values.id_9d)

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
#all_values.id$TF<-factor(all_values.id$TF, levels = levels(all_values.id$TF)[c(1:15,16:47,48:62,63:71,72,73:88)])

ggplot(data = JASPAR.all_values.id_9d, aes(x = cluster, y = gene)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()+ggtitle("9d TFs")


#do it per cluster
DefaultAssay(mo_9d)<- "chromvar"
Idents(mo_9d)<-"seurat_clusters"

JASPAR.DA.id_9d <- FindAllMarkers(
  object = mo_9d,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

t<- c()
for(i in 1:length(JASPAR.DA.id_9d$gene)){
  t[i]<-mo_9d@assays$ATAC@motifs@motif.names[[JASPAR.DA.id_9d$gene[i]]]
}


JASPAR.DA.id_9d$TF <- t

#select significant TFS
top_tf_ids_9d<- JASPAR.DA.id_9d %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
JASPAR.all_values.id_9d<- JASPAR.DA.id_9d[which(JASPAR.DA.id_9d$gene %in% c(top_tf_ids_9d$gene)),]

JASPAR.all_values.id_9d$cluster<-factor(JASPAR.all_values.id_9d$cluster, levels = c("1", "2", "3","4","5","6","7","8","9","10","11","12","13","14","15"))
JASPAR.all_values.id_9d<-JASPAR.all_values.id_9d %>% mutate(cluster=factor(cluster, levels = c("1", "2", "3","4","5","6","7","8","9","10","11","12","13","14","15"))) 
JASPAR.all_values.id_9d<-JASPAR.all_values.id_9d %>% arrange(cluster, desc(avg_diff))
JASPAR.all_values.id_9d$gene<-factor(JASPAR.all_values.id_9d$gene, levels =rev(unique(JASPAR.all_values.id_9d$gene)))
#Removes NA
JASPAR.all_values.id_9d <- na.omit(JASPAR.all_values.id_9d)

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
#all_values.id$TF<-factor(all_values.id$TF, levels = levels(all_values.id$TF)[c(1:15,16:47,48:62,63:71,72,73:88)])

ggplot(data = JASPAR.all_values.id_9d, aes(x = cluster, y = gene)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()+ggtitle("9d TFs")
```

#4.5d TF Dot Plots using JASPAR
```{r}
DefaultAssay(mo_4.5d)<- "chromvar"
Idents(mo_4.5d)<-"id"

JASPAR.DA.id_4.5d <- FindAllMarkers(
  object = mo_4.5d,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

t<- c()
for(i in 1:length(JASPAR.DA.id_4.5d$gene)){
  t[i]<-mo_4.5d@assays$ATAC@motifs@motif.names[[JASPAR.DA.id_4.5d$gene[i]]]
}


JASPAR.DA.id_4.5d$TF <- t

#select significant TFS
top_tf_ids_4.5d<- JASPAR.DA.id_4.5d %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
JASPAR.all_values.id_4.5d<- JASPAR.DA.id_4.5d[which(JASPAR.DA.id_4.5d$gene %in% c(top_tf_ids_4.5d$gene)),]

JASPAR.all_values.id_4.5d$cluster<-factor(JASPAR.all_values.id_4.5d$cluster, levels = c("BEAD_4.5d", "LD_4.5d", "HD_4.5d","3-27_4.5d","3-28_4.5d"))
JASPAR.all_values.id_4.5d<-JASPAR.all_values.id_4.5d %>% mutate(cluster=factor(cluster, levels = c("BEAD_4.5d", "LD_4.5d", "HD_4.5d","3-27_4.5d","3-28_4.5d"))) 
JASPAR.all_values.id_4.5d<-JASPAR.all_values.id_4.5d %>% arrange(cluster, desc(avg_diff))
JASPAR.all_values.id_4.5d$gene<-factor(JASPAR.all_values.id_4.5d$gene, levels =rev(unique(JASPAR.all_values.id_4.5d$gene)))
#Removes NA
JASPAR.all_values.id_4.5d <- na.omit(JASPAR.all_values.id_4.5d)

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
#all_values.id$TF<-factor(all_values.id$TF, levels = levels(all_values.id$TF)[c(1:15,16:47,48:62,63:71,72,73:88)])

ggplot(data = JASPAR.all_values.id_4.5d, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()+ggtitle("4.5d TFs per group")+labs(x = "group")

#PER CLUSTER
DefaultAssay(mo_4.5d)<- "chromvar"
Idents(mo_4.5d)<-"seurat_clusters"

JASPAR.DA.id_4.5d <- FindAllMarkers(
  object = mo_4.5d,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

t<- c()
for(i in 1:length(JASPAR.DA.id_4.5d$gene)){
  t[i]<-mo_4.5d@assays$ATAC@motifs@motif.names[[JASPAR.DA.id_4.5d$gene[i]]]
}


JASPAR.DA.id_4.5d$TF <- t

#select significant TFS
top_tf_ids_4.5d<- JASPAR.DA.id_4.5d %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
JASPAR.all_values.id_4.5d<- JASPAR.DA.id_4.5d[which(JASPAR.DA.id_4.5d$gene %in% c(top_tf_ids_4.5d$gene)),]

JASPAR.all_values.id_4.5d$cluster<-factor(JASPAR.all_values.id_4.5d$cluster, levels = c("2", "7", "5","6","3","4","1","0"))
JASPAR.all_values.id_4.5d<-JASPAR.all_values.id_4.5d %>% mutate(cluster=factor(cluster, levels = c("2", "7", "5","6","3","4","1","0"))) 
JASPAR.all_values.id_4.5d<-JASPAR.all_values.id_4.5d %>% arrange(cluster, desc(avg_diff))
JASPAR.all_values.id_4.5d$gene<-factor(JASPAR.all_values.id_4.5d$gene, levels =rev(unique(JASPAR.all_values.id_4.5d$gene)))
#Removes NA
JASPAR.all_values.id_4.5d <- na.omit(JASPAR.all_values.id_4.5d)

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
#all_values.id$TF<-factor(all_values.id$TF, levels = levels(all_values.id$TF)[c(1:15,16:47,48:62,63:71,72,73:88)])

#TF Dot Plot
ggplot(data = JASPAR.all_values.id_4.5d, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()+ggtitle("4.5d TFs per cluster")

#-----------------------------------------------------------------------------------------------------------------------------------------------------------------------

#Volcano plot comparing TFs in cluster 5 and 6
DefaultAssay(mo_4.5d) <- "chromvar"
markers1<-FindMarkers(mo_4.5d, group.by = "seurat_clusters", ident.1 = "5", ident.2 = "6",logfc.threshold = 0.1)

#get rid of lnf values
markers1$p_val_adj[markers1$p_val_adj == 0]<-1e-300

markers1$cluster<-"5"
markers1$cluster[markers1$avg_log2FC < 0] <- "6"

markers1$neg_log_pval <- -log10(markers1$p_val_adj)

markers1$gene<-rownames(markers1)

t<- c()
for(i in 1:length(markers1$gene)){
  t[i]<-mo_4.5d@assays$ATAC@motifs@motif.names[[markers1$gene[i]]]
}


# Load necessary libraries
library(TFBSTools)
library(rtracklayer)

# Import JASPAR data
jaspar_db <- getMatrixSet(JASPAR2020, opts=list(species=9606, all_versions=FALSE))

# Map JASPAR IDs to gene names
get_gene_name <- function(jaspar_id) {
  if (jaspar_id %in% names(jaspar_db)) {
    return(jaspar_db[[jaspar_id]]@name)
  } else {
    return(NA)
  }
}

markers1$gene_name <- sapply(markers1$gene, get_gene_name)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

options(ggrepel.max.overlaps = 50)


ggplot(markers1, aes(x = avg_log2FC, y = neg_log_pval, color = cluster))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene_name), min.segment.length = 0)+scale_color_manual(values = c("#fce503", "#8a9fd1"))+theme_ArchR()+ggtitle("4.5d Cluster 6 vs Cluster 5 TFs Volcano")


write.csv(markers1, file.path(RES_DIR, "4.5d_TFs_clusters_5v6.csv"))


library(ggplot2)
library(ggrepel)

# Sort and select top 30 genes per cluster
top_genes_5 <- markers1[markers1$cluster == "5", ] %>%
  arrange(p_val_adj, abs(avg_log2FC)) %>%
  head(30)

top_genes_6 <- markers1[markers1$cluster == "6", ] %>%
  arrange(p_val_adj, abs(avg_log2FC)) %>%
  head(30)

# Combine the top genes
top_genes <- rbind(top_genes_5, top_genes_6)

# Create the volcano plot
ggplot(markers1, aes(x = avg_log2FC, y = neg_log_pval, color = cluster)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = -0.2, linetype = "dashed", color = "gray80") +
  geom_vline(xintercept = 0.2, linetype = "dashed", color = "gray80") +
  geom_point() +
  geom_text_repel(data = top_genes, 
                  aes(label = gene_name), 
                  color = "black", 
                  size = 3, 
                  min.segment.length = 0) +
  scale_color_manual(values = c("#fce503", "#8a9fd1")) +
  theme_ArchR() +
  ggtitle("4.5d Cluster 6 vs Cluster 5 TFs Volcano")


```


#48hr TF Dot Plots using JASPAR
```{r}
DefaultAssay(mo_48hr)<- "chromvar"
rownames(mo_48hr@assays$chromvar)
FeaturePlot(mo_48hr, features = "TCF7")
FeaturePlot(mo_48hr,features = c("TCF7"), reduction = "wnn.umap")

Idents(mo_48hr)<-"id"

JASPAR.DA.id_48hr <- FindAllMarkers(
  object = mo_48hr,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

t<- c()
for(i in 1:length(JASPAR.DA.id_48hr$gene)){
  t[i]<-mo_48hr@assays$ATAC@motifs@motif.names[[JASPAR.DA.id_48hr$gene[i]]]
}


JASPAR.DA.id_48hr$TF <- t

#select significant TFS
top_tf_ids_48hr<- JASPAR.DA.id_48hr %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
JASPAR.all_values.id_48hr<- JASPAR.DA.id_48hr[which(JASPAR.DA.id_48hr$gene %in% c(top_tf_ids_48hr$gene)),]

JASPAR.all_values.id_48hr$cluster<-factor(JASPAR.all_values.id_48hr$cluster, levels = c("BEAD_48hr", "LD_48hr", "HD_48hr","3-27_48hr","3-28_48hr"))
JASPAR.all_values.id_48hr<-JASPAR.all_values.id_48hr %>% mutate(cluster=factor(cluster, levels = c("BEAD_48hr", "LD_48hr", "HD_48hr","3-27_48hr","3-28_48hr"))) 
JASPAR.all_values.id_48hr<-JASPAR.all_values.id_48hr %>% arrange(cluster, desc(avg_diff))
JASPAR.all_values.id_48hr$gene<-factor(JASPAR.all_values.id_48hr$gene, levels =rev(unique(JASPAR.all_values.id_48hr$gene)))
#Removes NA
JASPAR.all_values.id_48hr <- na.omit(JASPAR.all_values.id_48hr)

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
#all_values.id$TF<-factor(all_values.id$TF, levels = levels(all_values.id$TF)[c(1:15,16:47,48:62,63:71,72,73:88)])

ggplot(data = JASPAR.all_values.id_48hr, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()+ggtitle("48hr TFs per group")+labs(x = "group")

#PER CLUSTER
DefaultAssay(mo_48hr)<- "chromvar"
Idents(mo_48hr)<-"seurat_clusters"

JASPAR.DA.id_48hr <- FindAllMarkers(
  object = mo_48hr,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

t<- c()
for(i in 1:length(JASPAR.DA.id_48hr$gene)){
  t[i]<-mo_48hr@assays$ATAC@motifs@motif.names[[JASPAR.DA.id_48hr$gene[i]]]
}


JASPAR.DA.id_48hr$TF <- t

#select significant TFS
top_tf_ids_48hr<- JASPAR.DA.id_48hr %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)


#get unique pathways
#get values from other groups of top pathways
JASPAR.all_values.id_48hr<- JASPAR.DA.id_48hr[which(JASPAR.DA.id_48hr$gene %in% c(top_tf_ids_48hr$gene)),]

JASPAR.all_values.id_48hr$cluster<-factor(JASPAR.all_values.id_48hr$cluster, levels = c("3", "0", "4","2","5","1"))
JASPAR.all_values.id_48hr<-JASPAR.all_values.id_48hr %>% mutate(cluster=factor(cluster, levels = c("3", "0", "4","2","5","1"))) 
JASPAR.all_values.id_48hr<-JASPAR.all_values.id_48hr %>% arrange(cluster, desc(avg_diff))
JASPAR.all_values.id_48hr$gene<-factor(JASPAR.all_values.id_48hr$gene, levels =rev(unique(JASPAR.all_values.id_48hr$gene)))
#Removes NA
JASPAR.all_values.id_48hr <- na.omit(JASPAR.all_values.id_48hr)

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
#all_values.id$TF<-factor(all_values.id$TF, levels = levels(all_values.id$TF)[c(1:15,16:47,48:62,63:71,72,73:88)])

ggplot(data = JASPAR.all_values.id_48hr, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()+ggtitle("48hr TFs per cluster")

```

#Saving CSVs of TFs
```{r}
write.csv(top_tf_ids_48hr, file.path(RES_DIR, "top_TFs_48hr.csv"))
write.csv(top_tf_ids_4.5d, file.path(RES_DIR, "top_TFs_4'5d.csv"))
write.csv(top_tf_ids_9d, file.path(RES_DIR, "top_TFs_9d.csv"))
```

#Dot plots for motifs across each group (dot plot per group)
```{r DA motifs across exp groups DOT PLOT per group}
tfs<-levels(all_values$TF)

bead<-all_values[all_values$TF %in% tfs[1:15],]  %>% arrange(desc(avg_diff))
bead$TF<-factor(bead$TF, levels = rev(unique(bead$TF)))

c27<-all_values[all_values$TF %in% tfs[c(16:30)],]%>% arrange(desc(avg_diff))
c27$TF<-factor(c27$TF, levels = rev(unique(c27$TF)))

naive<-all_values[all_values$TF %in% tfs[31:45],]%>% arrange(desc(avg_diff))
naive$TF<-factor(naive$TF, levels = rev(unique(naive$TF)))

dim(naive)
dim(c27)
dim(bead)

DefaultAssay(MO)<-"chromvar"
FeaturePlot_scCustom(MO, features = "STAT1", reduction = "wnn.umap")

pdf("../figs/Motif data/Split versions_cisbp/naive_motf_acc_dot_plot.pdf", width = 2.75, height = 4.5)
ggplot(data = naive, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()
dev.off()

pdf("../figs/Motif data/Split versions_cisbp/cd27_motf_acc_dot_plot.pdf", width = 2.75, height = 4.5)
ggplot(data = c27, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()
dev.off()

pdf("../figs/Motif data/Split versions_cisbp/bead_motf_acc_dot_plot.pdf", width = 2.75, height = 4.5)
ggplot(data = bead, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()
dev.off()
```


#Violin plot for global accessibility, heatmap
```{r}
#9d
#VlnPlot(mo_9d, group.by = "id",  features = "log_ATAC", cols =c("Naive" = "gray75", "24hr_3/27" = "#008080", "24hr_Beads" = "Red"), pt.size = 0)
VlnPlot(mo_9d, group.by = "id",  features = "log_ATAC", cols = c("#186272","#f7ba97","red", "#5b278c", "#bd7ceb", "gray"), pt.size = 0)
VlnPlot(mo_9d_no_naive, group.by = "id",  features = "log_ATAC", cols = c("#186272","#f7ba97","red", "#5b278c", "#bd7ceb"), pt.size = 0)

#6x9
VlnPlot(mo_9d, group.by = "id",  features = "log_ATAC", cols = c("red","#bd7ceb","#5b278c", "#186272", "#f7ba97", "gray"), pt.size = 0)
mo_9d$id <- factor(mo_9d$id, levels = c("BEAD_9d", "LD_9d","HD_9d","3-27_9d","3-28_9d","Naive"))



data$Treatment <- factor(data$Treatment, levels=c("Y", "X", "Z"))

atac_mat<-MO@assays$ATAC@counts %>% as.matrix()

dim(atac_mat)
FindTopFeatures(MO, min.cutoff = 'q0') %>% TopFeatures()

atac_mat <-atac_mat[MO@assays$ATAC@var.features,]%>% as.matrix()
dim(atac_mat)

atac_mat[atac_mat > 0]<-1

colr = colorRamp2(
        seq(min(atac_mat), max(atac_mat), length = 2),
        c("white", "black"),
        space = "RGB")

group = MO$group
names(group) = colnames(MO)

column_ha = HeatmapAnnotation(
    group = MO$group, 
    col = list(group = c("Naive" = "gray75", "24hr_3/27" = "#008080", "24hr_Beads" = "Red")),
    annotation_name_gp= gpar(fontsize = 25),
    #border = T,
    annotation_legend_param = list(
        group = list(
            title = "group", 
            title_gp = gpar(fontsize = 25),
            labels_gp = gpar(fontsize = 20),
            border = T
        )
    )
)

# atac_mat<-AverageExpression(MO, assays = "ATAC", group.by = "group")

pdf(file.path(FIG_DIR, "Global Accessibility", "atac_heatmap_group.pdf"), width =6, height = 6)
ht = Heatmap(
    atac_mat,
    col= colr,
    use_raster = F,
    
    #anno
    top_annotation = column_ha,
    
    #legend
    name = "Binarized\ncounts",
    heatmap_legend_param = list(
        color_bar = "discrete",
        title_gp = gpar(fontsize = 25),
        labels_gp = gpar(fontsize = 20),
        at = 0:1,
        border = T
    ),
    
    #labels
    show_column_names = FALSE,
    show_row_names = FALSE,
    column_title = "", 
    row_title = "Top 2000 peaks",
    row_title_gp = gpar(fontsize = 25),
    border = TRUE,
    show_row_dend = F,
    
    #clustering
    clustering_distance_rows= "pearson",
    clustering_distance_columns= "pearson",
    clustering_method_rows = "ward.D2" ,
    clustering_method_columns="ward.D2",
    cluster_columns = cluster_within_group(atac_mat,group)
)
ht
dev.off()
```


#Make function that does volcano plot, (not specific genes annotated)
#need to add proper colors per treatment group
```{r}
#usage
#chromvar_volcano(seurat object, group_by[string], ident1[string], ident2[string], id_name1[string],id_name2[string], plot_title[string], pdf_name[string],csv_name[string])
chromvar_volcano <- function(seu_obj, group_by, ident1, ident2, id_name1, id_name2, plot_title, pdf_name, csv_name) {
  
  DefaultAssay(seu_obj) <- "chromvar"
  
  if (id_name1 == "3/27") {color1 = "#186272" }
  if (id_name1 == "3/28") {color1 = "#f8ba97" }
  if (id_name1 == "HD") {color1 = "#5b278c" }
  if (id_name1 == "LD") {color1 = "#bd7ceb" }
  if (id_name1 == "BEAD") {color1 = "#ff0000" }
  
  if (id_name2 == "3/27") {color2 = "#186272" }
  if (id_name2 == "3/28") {color2 = "#f8ba97" }
  if (id_name2 == "HD") {color2 = "#5b278c" }
  if (id_name2 == "LD") {color2 = "#bd7ceb" }
  if (id_name2 == "BEAD") {color2 = "#ff0000" }
  
  chromvar_markers<-FindMarkers(seu_obj, group.by = group_by, ident.1 = ident1, ident.2 = ident2, logfc.threshold = 0.1)
  
  chromvar_markers$gene<-rownames(chromvar_markers)
  
  t<- c()
  for(i in 1:length(chromvar_markers$gene)){
    t[i]<-seu_obj@assays$ATAC@motifs@motif.names[[chromvar_markers$gene[i]]]
  }

  chromvar_markers$TF <- t

  #get rid of lnf values
  chromvar_markers$p_val_adj[chromvar_markers$p_val_adj == 0]<-1e-300
  
  chromvar_markers$id<-id_name1
  chromvar_markers$id[chromvar_markers$avg_log2FC < 0] <- id_name2
  
  chromvar_markers$neg_log_pval <- -log10(chromvar_markers$p_val_adj)
  
  
  
  top_genes_1 <- chromvar_markers[chromvar_markers$id == id_name1, ] %>%
  arrange(p_val_adj, abs(avg_log2FC)) %>%
  head(30)

  top_genes_2 <- chromvar_markers[chromvar_markers$id == id_name2, ] %>%
  arrange(p_val_adj, abs(avg_log2FC)) %>%
  head(30)

# Combine the top genes
  top_genes <- rbind(top_genes_1, top_genes_2)

# Create the volcano plot
  myplot1<- ggplot(chromvar_markers, aes(x = avg_log2FC, y = neg_log_pval, color = id)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = -0.2, linetype = "dashed", color = "gray80") +
  geom_vline(xintercept = 0.2, linetype = "dashed", color = "gray80") +
  geom_point() +
  geom_text_repel(data = top_genes, 
                  aes(label = TF), 
                  color = "black", 
                  size = 3, 
                  min.segment.length = 0) +
  scale_color_manual(values = c(color1, color2)) +
  theme_ArchR() +
  ggtitle(plot_title)
  
  #chromvar_markers$gene<-rownames(chromvar_markers)
  
  pdf(file.path(FIG_DIR,"full_timecourse","TF_Volcano_Plots","TOP_30",pdf_name), width = 12, height =8)
  print(myplot1)     # Plot 1 --> in the first page of PDF
  dev.off()
  
  write.csv(chromvar_markers, file.path(RES_DIR,"TFs","TOP_30",csv_name))
}

```

#testing
```{r}
DefaultAssay(mo_48hr) <- "chromvar"

chromvar_markers<-FindMarkers(mo_48hr, group.by = "id", ident.1 = "3-27_48hr", ident.2 = "3-28_48hr", logfc.threshold = 0.1)

chromvar_markers$gene<-rownames(chromvar_markers)

t<- c()
for(i in 1:length(chromvar_markers$gene)){
  t[i]<-mo_48hr@assays$ATAC@motifs@motif.names[[chromvar_markers$gene[i]]]
}
chromvar_markers$TF <- t

#get rid of lnf values
chromvar_markers$p_val_adj[chromvar_markers$p_val_adj == 0]<-1e-300

chromvar_markers$id<-"3/27"
chromvar_markers$id[chromvar_markers$avg_log2FC < 0] <- "3/28"

chromvar_markers$neg_log_pval <- -log10(chromvar_markers$p_val_adj)



top_genes_1 <- chromvar_markers[chromvar_markers$id == "3/27", ] %>%
arrange(p_val_adj, abs(avg_log2FC)) %>%
head(30)

top_genes_2 <- chromvar_markers[chromvar_markers$id == "3/28", ] %>%
arrange(p_val_adj, abs(avg_log2FC)) %>%
head(30)

# Combine the top genes
top_genes <- rbind(top_genes_1, top_genes_2)

# Create the volcano plot
myplot1<- ggplot(chromvar_markers, aes(x = avg_log2FC, y = neg_log_pval, color = id)) +
geom_vline(xintercept = 0) +
geom_vline(xintercept = -0.2, linetype = "dashed", color = "gray80") +
geom_vline(xintercept = 0.2, linetype = "dashed", color = "gray80") +
geom_point() +
geom_text_repel(data = top_genes, 
                aes(label = TF), 
                color = "black", 
                size = 3, 
                min.segment.length = 0) +
scale_color_manual(values = c(color1, color2)) +
theme_ArchR() +
ggtitle("48hr 3/27 vs 3/28 Chromvar Volcano")

```


#volcano plots for every combination possible
```{r}
#48hr
chromvar_volcano(mo_48hr, "id", "3-27_48hr", "3-28_48hr", "3/27","3/28", "48hr 3/27 vs 3/28 Chromvar Volcano", "[48hr]3-27_vs_3-28_chromvar_volcano.pdf","[48hr]3-27_vs_3-28_chromvar_markers.csv")
chromvar_volcano(mo_48hr, "id", "3-27_48hr", "BEAD_48hr", "3/27","BEAD", "48hr 3/27 vs BEADs Chromvar Volcano", "[48hr]3-27_vs_BEADs_chromvar_volcano.pdf","[48hr]3-27_vs_BEADs_chromvar_markers.csv")
chromvar_volcano(mo_48hr, "id", "3-27_48hr", "HD_48hr", "3/27","HD", "48hr 3/27 vs HD Chromvar Volcano", "[48hr]3-27_vs_HD_chromvar_volcano.pdf","[48hr]3-27_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_48hr, "id", "3-27_48hr", "LD_48hr", "3/27","LD", "48hr 3/27 vs LD Chromvar Volcano", "[48hr]3-27_vs_LD_chromvar_volcano.pdf","[48hr]3-27_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_48hr, "id", "3-28_48hr", "BEAD_48hr", "3/28","BEAD", "48hr 3/28 vs BEADs Chromvar Volcano", "[48hr]3-28_vs_BEADs_chromvar_volcano.pdf","[48hr]3-28_vs_BEADs_chromvar_markers.csv")
chromvar_volcano(mo_48hr, "id", "3-28_48hr", "HD_48hr", "3/28","HD", "48hr 3/28 vs HD Chromvar Volcano", "[48hr]3-28_vs_HD_chromvar_volcano.pdf","[48hr]3-28_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_48hr, "id", "3-28_48hr", "LD_48hr", "3/28","LD", "48hr 3/28 vs LD Chromvar Volcano", "[48hr]3-28_vs_LD_chromvar_volcano.pdf","[48hr]3-28_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_48hr, "id", "BEAD_48hr", "HD_48hr", "BEAD","HD", "48hr BEADs vs HD Chromvar Volcano", "[48hr]BEADs_vs_HD_chromvar_volcano.pdf","[48hr]BEADs_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_48hr, "id", "BEAD_48hr", "LD_48hr", "BEAD","LD", "48hr BEADs vs LD Chromvar Volcano", "[48hr]BEADs_vs_LD_chromvar_volcano.pdf","[48hr]BEADs_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_48hr, "id", "HD_48hr", "LD_48hr", "HD","LD", "48hr HD vs LD Chromvar Volcano", "[48hr]HD_vs_LD_chromvar_volcano.pdf","[48hr]HD_vs_LD_chromvar_markers.csv")


#4.5d
chromvar_volcano(mo_4.5d, "id", "3-27_4.5d", "3-28_4.5d", "3/27","3/28", "4.5d 3/27 vs 3/28 Chromvar Volcano", "[4.5d]3-27_vs_3-28_chromvar_volcano.pdf","[4.5d]3-27_vs_3-28_chromvar_markers.csv")
chromvar_volcano(mo_4.5d, "id", "3-27_4.5d", "BEAD_4.5d", "3/27","BEAD", "4.5d 3/27 vs BEADs Chromvar Volcano", "[4.5d]3-27_vs_BEADs_chromvar_volcano.pdf","[4.5d]3-27_vs_BEADs_chromvar_markers.csv")
chromvar_volcano(mo_4.5d, "id", "3-27_4.5d", "HD_4.5d", "3/27","HD", "4.5d 3/27 vs HD Chromvar Volcano", "[4.5d]3-27_vs_HD_chromvar_volcano.pdf","[4.5d]3-27_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_4.5d, "id", "3-27_4.5d", "LD_4.5d", "3/27","LD", "4.5d 3/27 vs LD Chromvar Volcano", "[4.5d]3-27_vs_LD_chromvar_volcano.pdf","[4.5d]3-27_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_4.5d, "id", "3-28_4.5d", "BEAD_4.5d", "3/28","BEAD", "4.5d 3/28 vs BEADs Chromvar Volcano", "[4.5d]3-28_vs_BEADs_chromvar_volcano.pdf","[4.5d]3-28_vs_BEADs_chromvar_markers.csv")
chromvar_volcano(mo_4.5d, "id", "3-28_4.5d", "HD_4.5d", "3/28","HD", "4.5d 3/28 vs HD Chromvar Volcano", "[4.5d]3-28_vs_HD_chromvar_volcano.pdf","[4.5d]3-28_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_4.5d, "id", "3-28_4.5d", "LD_4.5d", "3/28","LD", "4.5d 3/28 vs LD Chromvar Volcano", "[4.5d]3-28_vs_LD_chromvar_volcano.pdf","[4.5d]3-28_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_4.5d, "id", "BEAD_4.5d", "HD_4.5d", "BEAD","HD", "4.5d BEADs vs HD Chromvar Volcano", "[4.5d]BEADs_vs_HD_chromvar_volcano.pdf","[4.5d]BEADs_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_4.5d, "id", "BEAD_4.5d", "LD_4.5d", "BEAD","LD", "4.5d BEADs vs LD Chromvar Volcano", "[4.5d]BEADs_vs_LD_chromvar_volcano.pdf","[4.5d]BEADs_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_4.5d, "id", "HD_4.5d", "LD_4.5d", "HD","LD", "4.5d HD vs LD Chromvar Volcano", "[4.5d]HD_vs_LD_chromvar_volcano.pdf","[4.5d]HD_vs_LD_chromvar_markers.csv")

#9d
chromvar_volcano(mo_9d, "id", "3-27_9d", "3-28_9d", "3/27","3/28", "9d 3/27 vs 3/28 Chromvar Volcano", "[9d]3-27_vs_3-28_chromvar_volcano.pdf","[9d]3-27_vs_3-28_chromvar_markers.csv")
chromvar_volcano(mo_9d, "id", "3-27_9d", "BEAD_9d", "3/27","BEAD", "9d 3/27 vs BEADs Chromvar Volcano", "[9d]3-27_vs_BEADs_chromvar_volcano.pdf","[9d]3-27_vs_BEADs_chromvar_markers.csv")
chromvar_volcano(mo_9d, "id", "3-27_9d", "HD_9d", "3/27","HD", "9d 3/27 vs HD Chromvar Volcano", "[9d]3-27_vs_HD_chromvar_volcano.pdf","[9d]3-27_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_9d, "id", "3-27_9d", "LD_9d", "3/27","LD", "9d 3/27 vs LD Chromvar Volcano", "[9d]3-27_vs_LD_chromvar_volcano.pdf","[9d]3-27_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_9d, "id", "3-28_9d", "BEAD_9d", "3/28","BEAD", "9d 3/28 vs BEADs Chromvar Volcano", "[9d]3-28_vs_BEADs_chromvar_volcano.pdf","[9d]3-28_vs_BEADs_chromvar_markers.csv")
chromvar_volcano(mo_9d, "id", "3-28_9d", "HD_9d", "3/28","HD", "9d 3/28 vs HD Chromvar Volcano", "[9d]3-28_vs_HD_chromvar_volcano.pdf","[9d]3-28_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_9d, "id", "3-28_9d", "LD_9d", "3/28","LD", "9d 3/28 vs LD Chromvar Volcano", "[9d]3-28_vs_LD_chromvar_volcano.pdf","[9d]3-28_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_9d, "id", "BEAD_9d", "HD_9d", "BEAD","HD", "9d BEADs vs HD Chromvar Volcano", "[9d]BEADs_vs_HD_chromvar_volcano.pdf","[9d]BEADs_vs_HD_chromvar_markers.csv")
chromvar_volcano(mo_9d, "id", "BEAD_9d", "LD_9d", "BEAD","LD", "9d BEADs vs LD Chromvar Volcano", "[9d]BEADs_vs_LD_chromvar_volcano.pdf","[9d]BEADs_vs_LD_chromvar_markers.csv")

chromvar_volcano(mo_9d, "id", "HD_9d", "LD_9d", "HD","LD", "9d HD vs LD Chromvar Volcano", "[9d]HD_vs_LD_chromvar_volcano.pdf","[9d]HD_vs_LD_chromvar_markers.csv")

```

#Volcano Plot 9d, 3/27 vs 3/28 (with specific genes annotated)
```{r}
DefaultAssay(mo_9d) <- "chromvar"
chromvar_markers_1<-FindMarkers(mo_9d, group.by = "id", ident.1 = "3-27_9d", ident.2 = "3-28_9d",logfc.threshold = 0.1)

#get rid of lnf values
chromvar_markers_1$p_val_adj[chromvar_markers_1$p_val_adj == 0]<-1e-300

chromvar_markers_1$id<-"3/27"
chromvar_markers_1$id[chromvar_markers_1$avg_log2FC < 0] <- "3/28"

chromvar_markers_1$neg_log_pval <- -log10(chromvar_markers_1$p_val_adj)

chromvar_markers_1$gene<-rownames(chromvar_markers_1)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

options(ggrepel.max.overlaps = 50)


ggplot(chromvar_markers_1, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#186272", "#ffb28d"))+theme_ArchR()+ggtitle("9d 3/27 vs 3/28 Chromvar Volcano")

dev.off()

write.csv(chromvar_markers_1, file.path(RES_DIR, "3-27_vs_3-28_chromvar_markers.csv"))


#manually annotate specific genes
chromvar_markers_1$genelabels <- ""
chromvar_markers_1$genelabels <- ifelse(chromvar_markers_1$gene == "RBAK" 
                              | chromvar_markers_1$gene == "ZNF37A"
                              | chromvar_markers_1$gene == "ZNF880"
                              | chromvar_markers_1$gene == "ZNF260"
                              | chromvar_markers_1$gene == "ZNF432"
                              | chromvar_markers_1$gene == "ZNF816"
                              | chromvar_markers_1$gene == "ZNF169"
                              | chromvar_markers_1$gene == "ZNF467"
                              | chromvar_markers_1$gene == "ZNF444"
                              | chromvar_markers_1$gene == "IKZF3"
                              | chromvar_markers_1$gene == "E2F6"
                              | chromvar_markers_1$gene == "E2F4"
                              | chromvar_markers_1$gene == "DNMT1"
                              | chromvar_markers_1$gene == "PRDM6"
                              | chromvar_markers_1$gene == "ZBTB17"
                              | chromvar_markers_1$gene == "GTF3A"
                              | chromvar_markers_1$gene == "SPI1"
                              | chromvar_markers_1$gene == "TCF7L2"
                              | chromvar_markers_1$gene == "RELA"
                              | chromvar_markers_1$gene == "CTCFL"
                              | chromvar_markers_1$gene == "OSR2"
                              | chromvar_markers_1$gene == "MECP2"
                              | chromvar_markers_1$gene == "PRDM5"
                              | chromvar_markers_1$gene == "TCF7L1"
                              | chromvar_markers_1$gene == "KLF17"
                              | chromvar_markers_1$gene == "TCF12"
                              | chromvar_markers_1$gene == "IRF1"
                              | chromvar_markers_1$gene == "KLF10"
                              | chromvar_markers_1$gene == "LEF1"
                              | chromvar_markers_1$gene == "CTCF"
                              | chromvar_markers_1$gene == "STAT2"
                              | chromvar_markers_1$gene == "TCF7"
                              | chromvar_markers_1$gene == "SOX4"
                              | chromvar_markers_1$gene == "MAZ"
                              | chromvar_markers_1$gene == "NFKB2"
                              | chromvar_markers_1$gene == "FOXG1"
                              | chromvar_markers_1$gene == "ASCL2"
                              | chromvar_markers_1$gene == "NFKB1"
                              | chromvar_markers_1$gene == "HINFP"
                              | chromvar_markers_1$gene == "SPIB"
                              | chromvar_markers_1$gene == "HDX"
                              | chromvar_markers_1$gene == "REL"
                              | chromvar_markers_1$gene == "ASCL1"
                              | chromvar_markers_1$gene == "IRF3"
                              | chromvar_markers_1$gene == "TFAP2E"
                              | chromvar_markers_1$gene == "E2F8"
                              | chromvar_markers_1$gene == "E2F7"
                              | chromvar_markers_1$gene == "SPIC"
                              | chromvar_markers_1$gene == "PRDM1"
                              | chromvar_markers_1$gene == "ZBTB1"
                              | chromvar_markers_1$gene == "SOX11"
                              | chromvar_markers_1$gene == "SOX9"
                              | chromvar_markers_1$gene == "ZBTB42"
                              | chromvar_markers_1$gene == "RELB"
                              | chromvar_markers_1$gene == "KLF2"
                              | chromvar_markers_1$gene == "RFX4"
                              | chromvar_markers_1$gene == "RFX2"
                              | chromvar_markers_1$gene == "RFX1"
                              | chromvar_markers_1$gene == "RFX3"
                              | chromvar_markers_1$gene == "RFX8"
                              | chromvar_markers_1$gene == "BACH2"
                              | chromvar_markers_1$gene == "BACH1"
                              | chromvar_markers_1$gene == "NFYB"
                              | chromvar_markers_1$gene == "FOSL1"
                              | chromvar_markers_1$gene == "NFYA"
                              | chromvar_markers_1$gene == "NFE2"
                              | chromvar_markers_1$gene == "FOS"
                              | chromvar_markers_1$gene == "JUND"
                              | chromvar_markers_1$gene == "NFYC"
                              | chromvar_markers_1$gene == "MITF"
                              | chromvar_markers_1$gene == "JUNB"
                              | chromvar_markers_1$gene == "FOSL2"
                              | chromvar_markers_1$gene == "JUN"
                              | chromvar_markers_1$gene == "JDP2"
                              | chromvar_markers_1$gene == "IRF2", TRUE, FALSE)


ggplot(chromvar_markers_1, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size = 3, label = ifelse(chromvar_markers_1$genelabels, chromvar_markers_1$gene, ""), min.segment.length = 0)+scale_color_manual(values = c("#186272", "#ffb28d"))+theme_ArchR()+ggtitle("9d 3/27 vs 3/28 Chromvar Volcano")


```

#Volcano Plot 9d, 3/27 vs BEADS
```{r}
DefaultAssay(mo_9d) <- "chromvar"
chromvar_markers_2<-FindMarkers(mo_9d, group.by = "id", ident.1 = "3-27_9d", ident.2 = "BEAD_9d",logfc.threshold = 0.1)

#get rid of lnf values
chromvar_markers_2$p_val_adj[chromvar_markers_2$p_val_adj == 0]<-1e-300

chromvar_markers_2$id<-"3/27"
chromvar_markers_2$id[chromvar_markers_2$avg_log2FC < 0] <- "BEAD"

chromvar_markers_2$neg_log_pval <- -log10(chromvar_markers_2$p_val_adj)

chromvar_markers_2$gene<-rownames(chromvar_markers_2)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

ggplot(chromvar_markers_2, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#186272", "red"))+theme_ArchR()+ggtitle("9d 3/27 vs BEADs Chromvar Volcano")

dev.off()
```

#Volcano Plot 9d, 3/28 vs BEADS
```{r}
DefaultAssay(mo_9d) <- "chromvar"
chromvar_markers_3<-FindMarkers(mo_9d, group.by = "id", ident.1 = "3-28_9d", ident.2 = "BEAD_9d",logfc.threshold = 0.1)

#get rid of lnf values
chromvar_markers_3$p_val_adj[chromvar_markers_3$p_val_adj == 0]<-1e-300

chromvar_markers_3$id<-"3/28"
chromvar_markers_3$id[chromvar_markers_3$avg_log2FC < 0] <- "BEAD"

chromvar_markers_3$neg_log_pval <- -log10(chromvar_markers_3$p_val_adj)

chromvar_markers_3$gene<-rownames(chromvar_markers_3)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

ggplot(chromvar_markers_3, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#ffb28d", "red"))+theme_ArchR()+ggtitle("9d 3/27 vs BEADs Chromvar Volcano")

dev.off()
```

# Volcano Plot comparing TF's in 4.5d clusters
```{r}
```


#figR functions
```{r}
plotfigRHeatmap<-function (figR.d, score.cut = 1, DORCs = NULL, TFs = NULL, ...) 
{
    message("Using absolute score cut-off of: ", score.cut, " ..\n")
    DORCsToKeep <- figR.d %>% dplyr::filter(abs(Score) >= score.cut) %>% 
        pull(DORC) %>% unique()
    TFsToKeep <- figR.d %>% dplyr::filter(abs(Score) >= score.cut) %>% 
        pull(Motif) %>% unique()
    if (!is.null(DORCs)) {
        if (!all(DORCs %in% figR.d$DORC)) 
            stop("One or more DORCs specified is not a valid DORC symbol found in the data.frame")
        DORCsToKeep <- intersect(DORCsToKeep, DORCs)
        TFsToKeep <- figR.d %>% dplyr::filter(abs(Score) >= score.cut & 
            DORC %in% DORCsToKeep) %>% pull(Motif) %>% unique()
    }
    if (!is.null(TFs)) {
        if (!all(TFs %in% figR.d$Motif)) 
            stop("One or more TFs specified is not a valid TF symbol found in the data.frame")
        TFsToKeep <- intersect(TFsToKeep, TFs)
        DORCsToKeep <- figR.d %>% dplyr::filter(abs(Score) >= 
            score.cut & Motif %in% TFsToKeep) %>% pull(DORC) %>% 
            unique()
    }
    net.d <- figR.d %>% dplyr::filter(DORC %in% DORCsToKeep & 
        Motif %in% TFsToKeep) %>% reshape2::dcast(DORC ~ Motif) %>% 
        tibble::column_to_rownames("DORC") %>% as.matrix()
    message("Plotting ", nrow(net.d), " DORCs x ", ncol(net.d), 
        "TFs\n")
    myCols <- circlize::colorRamp2(seq(-2, 2, length.out = 9), 
        colors = BuenColors::jdb_palette("solar_flare"))
    myHeat <- ComplexHeatmap::Heatmap(net.d, col = myCols,
         name = "Score", 
        border = TRUE, row_names_gp = gpar(fontsize = 15, fontface = "italic"), heatmap_width = unit(2, "npc"),
        ...)
    myHeat
}

rankDrivers<- function (figR.d, rankBy = c("meanScore", "nTargets"), myLabels = NULL, 
    score.cut = NULL, interactive = FALSE) 
{
    if (!rankBy %in% c("meanScore", "nTargets")) 
        stop("rankBy parameter has to be one of meanScore or nTargets to rank drivers using ..\n")
    if (rankBy %in% "meanScore") {
        message("Ranking TFs by mean regulation score across all DORCs ..\n")
        figR.summ <- figR.d %>% group_by(Motif) %>% dplyr::summarise(Score = mean(Score)) %>% 
            arrange(Score) %>% mutate(Motif = factor(Motif, 
            levels = as.character(Motif)))
        figR.summ$TF <- as.character(figR.summ$Motif)
        if (is.null(myLabels)) {
            figR.summ$TF[figR.summ$Score >= quantile(figR.summ$Score, 
                0.05) & figR.summ$Score <= quantile(figR.summ$Score, 
                0.95)] <- ""
        }
        else {
            figR.summ$TF[!figR.summ$TF %in% myLabels] <- ""
        }
        library(ggrepel)
    gAll <- ggplot(figR.summ,aes(x=Motif,y=Score,label=TF)) + 
      geom_bar(size=0.1,stat="identity",fill="darkorange",color=NA) + 
      theme_classic() + theme(axis.text.x = element_text(size = 7), axis.text = element_text(size = 7), axis.title = element_text(size = 20)) + 
      geom_hline(yintercept = 0) + labs(x="TF Motifs",y="Regulation Score")+coord_flip()
    
    gAll
    }
}

plot_drivers<-function (figR.d, marker, score.cut = 1, label = TRUE) 
{
    if (!marker %in% figR.d$DORC) 
        stop("Marker specified is not a valid DORC symbol found in the data.frame")
    d <- figR.d %>% dplyr::filter(DORC %in% marker) %>% mutate(isSig = ifelse(abs(Score) >= 
        score.cut, "Yes", "No"))
    if (label) {
        d$Label <- d$Motif
        d$Label[d$isSig %in% "No"] <- ""
    }
    else {
        d$Label <- ""
    }
    gScatter <- d %>% ggplot(aes(x = Corr.log10P, y = Enrichment.log10P, 
        color = isSig, label = Label)) + geom_hline(yintercept = 0, 
        color = "gray60", linetype = "dashed") + geom_vline(xintercept = 0, 
        color = "gray60", linetype = "dashed") + geom_point(size = 0.8) + 
        theme_classic() + scale_color_manual(values = c("gray66", 
        "firebrick3")) + scale_x_continuous(breaks = scales::pretty_breaks()) + 
        scale_y_continuous(breaks = scales::pretty_breaks()) + 
        labs(y = "Enrichment log10 P", x = "Correlation log10 P", 
            title = marker) + ylim(-ceiling(max(abs(d$Enrichment.log10P))), 
        ceiling(max(abs(d$Enrichment.log10P)))) + xlim(-ceiling(max(abs(d$Corr.log10P))), 
        ceiling(max(abs(d$Corr.log10P)))) + theme(legend.position = "none", 
        axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5, 
            face = "italic"), panel.background = element_rect(fill = NA)) + 
        geom_text(hjust = 1.1, fontface = "italic", color = "black", 
            size = 3)
    gScatter
}

```

#top regulated TFs for 48hrs
```{r}
#prepare multi ome data

#!!! Could be broken because object should be (peaks x cells), not (cells x peaks)!!!!
ATAC.se_48hr <- SummarizedExperiment(assays = list(counts = mo_48hr[["ATAC"]]@data) , rowRanges = mo_48hr[["ATAC"]]@ranges, colData = mo_48hr@meta.data)

#!!! Could be broken because matrix should be (genes x cells), not (cells x genes)!!!!
RNAmat_48hr<- mo_48hr[["SCT"]]@data %>% as.sparse()

#ran for a long time

#cisCor_48hr <- runGenePeakcorr(ATAC.se = ATAC.se_48hr,
#                           RNAmat = RNAmat_48hr,
#                           genome = "hg38", # Also supports mm10 and hg38
#                           nCores = detectCores(), 
#                           p.cut = NULL)

#write.csv(cisCor_48hr, file.path(RES_DIR,"cisCor_48hr.csv"))
cisCor_48hr<-read.csv(file.path(RES_DIR, "cisCor_48hr.csv"))
                    
# Filter peak-gene correlations by p-value                    
cisCor.filt_48hr <- cisCor_48hr %>% dplyr::filter(pvalZ <= 0.05)

# Determine DORC genes
dp_48hr <- cisCor.filt_48hr %>% dorcJPlot(cutoff=4, # Default
                                       returnGeneList = F, labelTop = 10, labelSize = 3, cleanLabels = T)


dorcGenes_48hr <-c("IKZF2","GBP5","TIGIT","CSF2","ADAM19","GIMAP7","TNFSF10","RGS1","IFNG","LINC-PINT", "SELL")

# Get DORC scores
dorcMat_48hr <- getDORCScores(ATAC.se_48hr ,dorcTab=cisCor.filt_48hr,geneList=dorcGenes_48hr,nCores=detectCores())

lsi_48hr <- mo_48hr@reductions$ATAC_UMAP@cell.embeddings %>% as.data.frame()

#trying lsi using ATAC coordinates instead
#lsi_48hr <- mo_48hr@reductions$ATAC_UMAP@cell.embeddings %>% as.data.frame()

# rownames(lsi)<- gsub("#", "_",rownames(lsi) )

lsi_48hr <- lsi_48hr[which(rownames(lsi_48hr) %in% colnames(dorcMat_48hr)),]
dorcMat_48hr <- dorcMat_48hr[,which(colnames(dorcMat_48hr) %in% rownames(lsi_48hr))]
ATAC.se_48hr <- ATAC.se_48hr[, which(colnames(ATAC.se_48hr) %in% rownames(lsi_48hr))]
RNAmat_48hr <- RNAmat_48hr[, which(colnames(RNAmat_48hr) %in% rownames(lsi_48hr))]


lsi.knn_48hr <- FNN::get.knn(lsi_48hr,k=2)$nn.index
rownames(lsi.knn_48hr) <- colnames(dorcMat_48hr)

# Smooth DORC scores (using cell KNNs)
dorcMat.smooth_48hr <- smoothScoresNN(NNmat=lsi.knn_48hr,mat=dorcMat_48hr,nCores=detectCores())

human_pwms_v3 <- readRDS("~/cisBP_human_pfms_2021.rds")
rnaMat.smoothed_48hr <- smoothScoresNN(NNmat = lsi.knn_48hr,mat = RNAmat_48hr,nCores = detectCores(),geneList=intersect(rownames(RNAmat_48hr),names(human_pwms_v3)))

fig.d_48hr <- runFigRGRN(ATAC.se=ATAC.se_48hr,
                    rnaMat=rnaMat.smoothed_48hr, # Smoothed RNA matrix using paired cell kNNs
                    dorcMat=dorcMat.smooth_48hr,
                    dorcTab=cisCor.filt_48hr,
                    genome="hg38",
                    dorcGenes=dorcGenes_48hr,
                    nCores=detectCores())
```

#top regulated TFs for 4.5d
```{r}
#prepare multi ome data
ATAC.se_4.5d <- SummarizedExperiment(assays = list(counts = mo_4.5d[["ATAC"]]@data) , rowRanges = mo_4.5d[["ATAC"]]@ranges, colData = mo_4.5d@meta.data)

RNAmat_4.5d<- mo_4.5d[["SCT"]]@data %>% as.sparse()

#ran for a long time
cisCor_4.5d <- runGenePeakcorr(ATAC.se = ATAC.se_4.5d,
                           RNAmat = RNAmat_4.5d,
                           genome = "hg38", # Also supports mm10 and hg38
                           nCores = detectCores(), 
                           p.cut = NULL)

write.csv(cisCor_4.5d, file.path(RES_DIR,"cisCor_4.5d.csv"))
#cisCor_48hr<-read.csv(file.path(RES_DIR, "cisCor_48hr.csv"))
                    
# Filter peak-gene correlations by p-value                    
cisCor.filt_48hr <- cisCor_48hr %>% dplyr::filter(pvalZ <= 0.05)

# Determine DORC genes
dp_48hr <- cisCor.filt_48hr %>% dorcJPlot(cutoff=4, # Default
                                       returnGeneList = F, labelTop = 10, labelSize = 3, cleanLabels = T)


dorcGenes_48hr <-c("IKZF2","GBP5","TIGIT","CSF2","ADAM19","GIMAP7","TNFSF10","RGS1","IFNG","LINC-PINT", "SELL")

# Get DORC scores
dorcMat_48hr <- getDORCScores(ATAC.se_48hr ,dorcTab=cisCor.filt_48hr,geneList=dorcGenes_48hr,nCores=detectCores())

lsi_48hr <- mo_48hr@reductions$ATAC_UMAP@cell.embeddings %>% as.data.frame()

#trying lsi using ATAC coordinates instead
#lsi_48hr <- mo_48hr@reductions$ATAC_UMAP@cell.embeddings %>% as.data.frame()

# rownames(lsi)<- gsub("#", "_",rownames(lsi) )

lsi_48hr <- lsi_48hr[which(rownames(lsi_48hr) %in% colnames(dorcMat_48hr)),]
dorcMat_48hr <- dorcMat_48hr[,which(colnames(dorcMat_48hr) %in% rownames(lsi_48hr))]
ATAC.se_48hr <- ATAC.se_48hr[, which(colnames(ATAC.se_48hr) %in% rownames(lsi_48hr))]
RNAmat_48hr <- RNAmat_48hr[, which(colnames(RNAmat_48hr) %in% rownames(lsi_48hr))]


lsi.knn_48hr <- FNN::get.knn(lsi_48hr,k=2)$nn.index
rownames(lsi.knn_48hr) <- colnames(dorcMat_48hr)

# Smooth DORC scores (using cell KNNs)
dorcMat.smooth_48hr <- smoothScoresNN(NNmat=lsi.knn_48hr,mat=dorcMat_48hr,nCores=detectCores())

human_pwms_v3 <- readRDS("~/cisBP_human_pfms_2021.rds")
rnaMat.smoothed_48hr <- smoothScoresNN(NNmat = lsi.knn_48hr,mat = RNAmat_48hr,nCores = detectCores(),geneList=intersect(rownames(RNAmat_48hr),names(human_pwms_v3)))

fig.d_48hr <- runFigRGRN(ATAC.se=ATAC.se_48hr,
                    rnaMat=rnaMat.smoothed_48hr, # Smoothed RNA matrix using paired cell kNNs
                    dorcMat=dorcMat.smooth_48hr,
                    dorcTab=cisCor.filt_48hr,
                    genome="hg38",
                    dorcGenes=dorcGenes_48hr,
                    nCores=detectCores())
```


#top regulated TFs for 9d
```{r}
#prepare multi ome data
ATAC.se_9d <- SummarizedExperiment(assays = list(counts = mo_9d[["ATAC"]]@data) , rowRanges = mo_9d[["ATAC"]]@ranges, colData = mo_9d@meta.data)

RNAmat_9d<- mo_9d[["SCT"]]@data %>% as.sparse()

#ran for a long time
cisCor_9d <- runGenePeakcorr(ATAC.se = ATAC.se_9d,
                           RNAmat = RNAmat_9d,
                           genome = "hg38", # Also supports mm10 and hg38
                           nCores = detectCores(), 
                           p.cut = NULL)

write.csv(cisCor_9d, file.path(RES_DIR,"cisCor_9d.csv"))
#cisCor_48hr<-read.csv(file.path(RES_DIR, "cisCor_48hr.csv"))
                    
# Filter peak-gene correlations by p-value                    
cisCor.filt_48hr <- cisCor_48hr %>% dplyr::filter(pvalZ <= 0.05)

# Determine DORC genes
dp_48hr <- cisCor.filt_48hr %>% dorcJPlot(cutoff=4, # Default
                                       returnGeneList = F, labelTop = 10, labelSize = 3, cleanLabels = T)


dorcGenes_48hr <-c("IKZF2","GBP5","TIGIT","CSF2","ADAM19","GIMAP7","TNFSF10","RGS1","IFNG","LINC-PINT", "SELL")

# Get DORC scores
dorcMat_48hr <- getDORCScores(ATAC.se_48hr ,dorcTab=cisCor.filt_48hr,geneList=dorcGenes_48hr,nCores=detectCores())

lsi_48hr <- mo_48hr@reductions$ATAC_UMAP@cell.embeddings %>% as.data.frame()

#trying lsi using ATAC coordinates instead
#lsi_48hr <- mo_48hr@reductions$ATAC_UMAP@cell.embeddings %>% as.data.frame()

# rownames(lsi)<- gsub("#", "_",rownames(lsi) )

lsi_48hr <- lsi_48hr[which(rownames(lsi_48hr) %in% colnames(dorcMat_48hr)),]
dorcMat_48hr <- dorcMat_48hr[,which(colnames(dorcMat_48hr) %in% rownames(lsi_48hr))]
ATAC.se_48hr <- ATAC.se_48hr[, which(colnames(ATAC.se_48hr) %in% rownames(lsi_48hr))]
RNAmat_48hr <- RNAmat_48hr[, which(colnames(RNAmat_48hr) %in% rownames(lsi_48hr))]


lsi.knn_48hr <- FNN::get.knn(lsi_48hr,k=2)$nn.index
rownames(lsi.knn_48hr) <- colnames(dorcMat_48hr)

# Smooth DORC scores (using cell KNNs)
dorcMat.smooth_48hr <- smoothScoresNN(NNmat=lsi.knn_48hr,mat=dorcMat_48hr,nCores=detectCores())

human_pwms_v3 <- readRDS("~/cisBP_human_pfms_2021.rds")
rnaMat.smoothed_48hr <- smoothScoresNN(NNmat = lsi.knn_48hr,mat = RNAmat_48hr,nCores = detectCores(),geneList=intersect(rownames(RNAmat_48hr),names(human_pwms_v3)))

fig.d_48hr <- runFigRGRN(ATAC.se=ATAC.se_48hr,
                    rnaMat=rnaMat.smoothed_48hr, # Smoothed RNA matrix using paired cell kNNs
                    dorcMat=dorcMat.smooth_48hr,
                    dorcTab=cisCor.filt_48hr,
                    genome="hg38",
                    dorcGenes=dorcGenes_48hr,
                    nCores=detectCores())
```
#=============================================================================================================================================================
#=============================================================================================================================================================
#=============================================================================================================================================================


#Exporting data to GEO
```{r}
sample2 <- mo_naive_24hr[,mo_naive_24hr$id != 'BEAD_24hr']
sample3 <- mo_naive_24hr[,mo_naive_24hr$id == 'BEAD_24hr']

#Sample 2
rna_matrix_2 <- sample2[["RNA"]]@counts
write.table(rna_matrix_2, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_28_2023/processed_data/S2_gene_expression_matrix.txt", sep="\t", row.names = TRUE, col.names=TRUE, quote=FALSE)
#write.csv(rna_matrix_2, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_28_2023/processed_data/S2_gene_expression_matrix.csv")

rna_matrix <- seurat_obj[["RNA"]]@counts
rna_matrix <- data.frame(Gene_Symbols = rownames(rna_matrix), rna_matrix)
write.table(rna_matrix, file="rna_expression_matrix_with_gene_symbols.txt", sep="\t", col.names=TRUE, quote=FALSE)


atac_matrix_2 <- sample2[["ATAC"]]@counts
write.table(atac_matrix_2, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_20_2023/processed_data/S2_atac_feature_matrix.txt", sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)

metadata_2 <- sample2@meta.data
write.table(metadata_2, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_20_2023/processed_data/S2_metadata.txt", sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)

#Sample 3
rna_matrix_3 <- sample3[["RNA"]]@counts
write.table(rna_matrix_3, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_20_2023/processed_data/S3_gene_expression_matrix.txt", sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)

atac_matrix_3 <- sample3[["ATAC"]]@counts
write.table(atac_matrix_3, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_20_2023/processed_data/S3_atac_feature_matrix.txt", sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)

metadata_3 <- sample3@meta.data
write.table(metadata_3, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_20_2023/processed_data/S3_metadata.txt", sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)



#HTO Matrix For Pool 1 (Sample 2)
library(Matrix)
library(data.table)

# Read the files
barcodes <- fread(file = "/fh/fast/furlan_s/experiments/timecourse_cd27/MO_cd27/HTO_count/CJMOP1_HTAL/read_count/barcodes.tsv.gz",header = F)
features <- fread(file = "/fh/fast/furlan_s/experiments/timecourse_cd27/MO_cd27/HTO_count/CJMOP1_HTAL/read_count/features.tsv.gz", header = F)
mtx <- as(readMM(file = "/fh/fast/furlan_s/experiments/timecourse_cd27/MO_cd27/HTO_count/CJMOP1_HTAL/read_count/matrix.mtx.gz"), "CsparseMatrix")


# Convert the matrix to a dense format and use data.table for easier handling
dt_mtx <- as.data.table(as.matrix(mtx))

#Set the rownames to features
features <- t(features)
rownames(dt_mtx) <- features

#Set colnames to barcodes
barcodes <- t(barcodes)
colnames(dt_mtx) <- barcodes


write.table(dt_mtx, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_28_2023/processed_data/S2_tsa_count_matrix.txt", sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE)

#HTO Matrix For Pool 2 (Sample 3)
library(Matrix)
library(data.table)

# Read the files
barcodes <- fread(file = "/fh/fast/furlan_s/experiments/timecourse_cd27/MO_cd27/HTO_count/CJMOP2_HTAL/read_count/barcodes.tsv.gz",header = F)
features <- fread(file = "/fh/fast/furlan_s/experiments/timecourse_cd27/MO_cd27/HTO_count/CJMOP2_HTAL/read_count/features.tsv.gz", header = F)
mtx <- as(readMM(file = "/fh/fast/furlan_s/experiments/timecourse_cd27/MO_cd27/HTO_count/CJMOP2_HTAL/read_count/matrix.mtx.gz"), "CsparseMatrix")


# Convert the matrix to a dense format and use data.table for easier handling
dt_mtx <- as.data.table(as.matrix(mtx))

#Set the rownames to features
features <- t(features)
rownames(dt_mtx) <- features

#Set colnames to barcodes
barcodes <- t(barcodes)
colnames(dt_mtx) <- barcodes


write.table(dt_mtx, file="/fh/fast/furlan_s/user/pmuthura/geo_uploads/C_Jaeger_9_28_2023/processed_data/S3_tsa_count_matrix.txt", sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE)
```


#ATAC UMAP for 24hr+naive
```{r reduce dimensions and run UMAP and plot}
mo_naive_24hr_void_3_28<-readRDS(file.path(CDS_DIR, "3-28_void_timecourse_MO_cd27_naive_24hr.rds"))

DefaultAssay(mo_naive_24hr) <- "ATAC"
mo_naive_24hr <- RunTFIDF(mo_naive_24hr)
mo_naive_24hr <- FindTopFeatures(mo_naive_24hr) #Find most frequently observed features
mo_naive_24hr<- RunSVD(mo_naive_24hr)
ElbowPlot(mo_naive_24hr,reduction = 'lsi')
mo_naive_24hr <- RunUMAP(mo_naive_24hr, reduction = 'lsi', dims = 2:50, reduction.name = "ATAC_UMAP") #running dimensional reduction

DimPlot(mo_naive_24hr,  group.by ="dataset", label=T, reduction = "ATAC_UMAP")+ggtitle("ATAC UMAP") #UMAP plot
DimPlot(mo_naive_24hr,  group.by ="id", label=T, reduction = "ATAC_UMAP",label.box = T,  cols = paletteDiscrete(values = levels(factor(mo_naive_24hr$id))))+ggtitle("ATAC UMAP") #UMAP plot
DimPlot(mo_naive_24hr,  group.by ="timepoint", label=T, reduction = "ATAC_UMAP",label.box = T,  cols = paletteDiscrete(values = levels(factor(mo$timepoint))))+ggtitle("ATAC UMAP") #UMAP plot

```

#RNA/SCT UMAP for 24hr+naive
```{r}
DefaultAssay(mo_naive_24hr) <- "RNA"
mo_naive_24hr<-FindVariableFeatures(mo_naive_24hr) #Identifies features that are outliers on a 'mean variability plot'.
mo_naive_24hr<-ScaleData(mo_naive_24hr) #Scale/Center data
mo_naive_24hr <- SCTransform(mo_naive_24hr, method = "glmGamPoi", vars.to.regress = "percent.mt") #A normalization method for single-cell UMI count data using a variance stabilizing transformation.
mo_naive_24hr <- RunPCA(mo_naive_24hr, features = VariableFeatures(mo_naive_24hr))
ElbowPlot(mo_naive_24hr, ndims = 30)
mo_naive_24hr<-RunUMAP(mo_naive_24hr, dims = 1:50, verbose = T, reduction.name = "SCT_UMAP") #UMAP for SCT method
DimPlot(mo_naive_24hr,  group.by ="id", label=T, reduction = "SCT_UMAP")+ggtitle("SCT UMAP")

```


#WNN Umap and Seurat Clustering for Naive and 24hr
```{r}
#mo_naive_24hr<-readRDS(file.path(CDS_DIR, "timecourse_MO_cd27_naive_24hr.rds"))

mo_naive_24hr <- FindMultiModalNeighbors(mo_naive_24hr, reduction.list = list("pca", "lsi"), dims.list = list(1:50, 2:50), weighted.nn.name = "wnn.nn")
mo_naive_24hr <- RunUMAP(mo_naive_24hr, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

#umap plots
DimPlot(mo_24hr, group.by = "id", reduction = "wnn.umap", cols = c("#186272", "#ffb28d", "red","#5b278c","#bd7ceb","gray"),label = F, label.box = F)+ggtitle("WNN UMAP")&NoAxes()
DimPlot(mo_naive_24hr, group.by = "timepoint", reduction = "wnn.umap", label = T)+ggtitle("WNN UMAP")
DimPlot(mo_naive_24hr,  group.by ="id", label=T, reduction = "wnn.umap",split.by ="timepoint",label.size = 3,label.box = T)+ggtitle("WNN UMAP")

#Using monocle for a nicer umap
cds<-seurat_to_monocle3(mo_naive_24hr, seu_rd = "wnn.umap")
cluster_cells(cds)
plot_cells(cds, color_cells_by = "id", cell_size = 0.7, label_cell_groups = F)+scale_color_manual(values = c("#186272", "red","#5b278c","#bd7ceb","gray"))+
  theme(legend.title=element_blank())&NoAxes()

#Feature Plots for marker gene expression
FeaturePlot(mo_naive_24hr,features = c("TCF7","SELL","LEF1","IL2","IL2RA","FOXP1","MYC","STAT1","TBX21","EOMES","BACH1","FOS","JUN","MALAT1","CCR7","PRMT2","CD96","IFNG"), reduction = "wnn.umap", order = T,cols = c("gray70", "#04578f", "#801b1b"))&NoAxes()
FeaturePlot(mo_naive_24hr,features = c("IKZF1","IKZF2"), reduction = "wnn.umap",cols = c("gray70", "#04578f", "#801b1b"), order = T)&NoAxes()

FeaturePlot(mo_naive_24hr,features = c("MKI67"), reduction = "wnn.umap",cols = c("gray70", "#04578f", "#801b1b"), order = T)&NoAxes()

#Seurat Clustering----------------------------------------------------------------------------------------------------------------------------------------------------------------
mo_naive_24hr <- FindClusters(mo_naive_24hr, resolution = 0.20, graph.name = "wsnn", random.seed = 123) #5 nice clusters


#Checking cluster plot
DimPlot(mo_naive_24hr, group.by = "seurat_clusters", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_naive_24hr$seurat_clusters)), set = "stallion2"))+ggtitle("Seurat clusters")&NoAxes()

#Subclustering cluster 0, then creating new metadata in main object based on this subset object
cluster0_data <- subset(mo_naive_24hr, subset = seurat_clusters == 0)
cluster0_data <- FindNeighbors(cluster0_data, dims = 1:10, k.param = 5)
cluster0_data <- FindClusters(cluster0_data, resolution = 0.08, graph.name = "wsnn", random.seed = 123)

mo_naive_24hr$sub_cluster_0 <- as.character(Idents(mo_naive_24hr))
mo_naive_24hr$sub_cluster_0[Cells(cluster0_data)] <- paste("c0",Idents(cluster0_data))

#checking cluster plot
DimPlot(mo_naive_24hr, group.by = "sub_cluster_0", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_naive_24hr$sub_cluster_0)), set = "stallion2"))+ggtitle("Seurat clusters")&NoAxes()

#subclustering cluster 4 then adding to the new metadata column made before
cluster4_data <- subset(mo_naive_24hr, subset = seurat_clusters == 4)
cluster4_data <- FindNeighbors(cluster4_data, dims = 1:10, k.param = 5)
cluster4_data <- FindClusters(cluster4_data, resolution = 0.15, graph.name = "wsnn", random.seed = 123)

mo_naive_24hr$sub_cluster_0[Cells(cluster4_data)] <- paste("c4",Idents(cluster4_data))

#renaming metadata to test for downstream stuff
mo_naive_24hr$test<-mo_naive_24hr$sub_cluster_0

DimPlot(mo_naive_24hr, group.by = "test", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_naive_24hr$sub_cluster_0)), set = "stallion2"))+ggtitle("Seurat clusters")&NoAxes()

#Merging and renaming clusters
mo_naive_24hr$test[mo_naive_24hr$test == "c4 1"]<-"1"
mo_naive_24hr$test[mo_naive_24hr$test == "one"]<-"six"
mo_naive_24hr$test[mo_naive_24hr$test == "c4 0"]<-"two" 
mo_naive_24hr$test[mo_naive_24hr$test == "3"]<-"three" 
mo_naive_24hr$test[mo_naive_24hr$test == "c0 0"]<-"four" 
mo_naive_24hr$test[mo_naive_24hr$test == "c0 1"]<-"five" 
mo_naive_24hr$test[mo_naive_24hr$test == "c"]<-"four" 
mo_naive_24hr$test[mo_naive_24hr$test == "2"]<-"one"

mo_naive_24hr$test[mo_naive_24hr$test == "one"] <- "1"
mo_naive_24hr$test[mo_naive_24hr$test == "two"] <- "2"
mo_naive_24hr$test[mo_naive_24hr$test == "three"] <- "3"
mo_naive_24hr$test[mo_naive_24hr$test == "four"] <- "4"
mo_naive_24hr$test[mo_naive_24hr$test == "five"] <- "5"
mo_naive_24hr$test[mo_naive_24hr$test == "six"] <- "6"



DimPlot(mo_naive_24hr, group.by = "test", reduction = "wnn.umap", label = T,label.box = T, cols =  ArchR::paletteDiscrete(values = levels(factor(mo_naive_24hr$test)), set = "stallion2"))+ggtitle("Seurat clusters")&NoAxes()

#WNN umap with custom colors
DimPlot(mo_naive_24hr, group.by = "test", reduction = "wnn.umap", label = F,label.box = F, cols = c("#332747","#56919e","#6bb2a4","#424b81","#bddbc2","#436c8e"))+ggtitle("Seurat Clusters")&NoAxes()

DimPlot(mo_naive_24hr, group.by = "test", reduction = "wnn.umap", label = F,label.box = F, cols = c("#581845","#900c3f","#ffc300","#ff5733","#c70039","#a742f5"))+ggtitle("Seurat Clusters")&NoAxes()

#5x7
#final choice:
DimPlot(mo_naive_24hr, group.by = "test", reduction = "wnn.umap", label = F,label.box = F, cols = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+ggtitle("Seurat Clusters")&NoAxes()

#0, 3, 1, 5, 2, 4

#5x7
#revised choice:
DimPlot(mo_naive_24hr, group.by = "test", reduction = "wnn.umap", label = F,label.box = F, cols = c("#000000","#17068b","#9915a0","#d7566c","#f79242","#f7e025"))+ggtitle("Seurat Clusters")&NoAxes()





#Renaming 24 hour id's so they can be called in the loop below without breaking
mo_naive_24hr$id[mo_naive_24hr$dataset == "CJ_MO_P1" & mo_naive_24hr$HTO_maxID == 'HT4-AGTAAGTTCAGCGTA']<- "3-27_24hr"
mo_naive_24hr$id[mo_naive_24hr$dataset == "CJ_MO_P1" & mo_naive_24hr$HTO_maxID == 'HT3-TTCCGCCTCTCTTTG']<- "HD_24hr"
mo_naive_24hr$id[mo_naive_24hr$dataset == "CJ_MO_P1" & mo_naive_24hr$HTO_maxID == 'HT2-TGATGGCCTATTGGG']<- "LD_24hr"
mo_naive_24hr$id[mo_naive_24hr$dataset == "CJ_MO_P1" & mo_naive_24hr$HTO_maxID == 'HT1-GTCAACTCTTTAGCG']<- "3-28_24hr"
mo_naive_24hr$id[mo_naive_24hr$dataset == "CJ_MO_P2" & mo_naive_24hr$HTO_maxID == 'HT8-CTCCTCTGCAATTAC']<- "BEAD_24hr"

#Color each cluster individually (in ID UMAP)
for (i in 1:length(levels(factor(mo_naive_24hr$id)))) {
  id_cols<-rep("gray80",length(levels(factor(mo_naive_24hr$id))))
  names(id_cols)<-levels(factor(mo_naive_24hr$id))
  id_cols[i]<-"red"
  pdf(paste0("/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/figs/",names(id_cols)[i],"_naive_24hr_umap_VOID_3_28.pdf"),width=10,height=10)
  print(DimPlot(mo_naive_24hr, group.by = "id", reduction = "wnn.umap", label = F, cols = id_cols, label.box = F)+NoAxes()+ggtitle("WNN UMAP"))
  dev.off()
  }

#STOPPED HERE!!!!!!!!!!!!!!
#Find Marker genes per cluster
Idents(mo_naive_24hr)<-"test"
mo_naive_24hr.markers <- FindAllMarkers(mo_naive_24hr, only.pos = FALSE,  logfc.threshold = 0.25) #This step runs for a very long time on large datasets

#changed from 8-->12
top_n<- mo_naive_24hr.markers %>%
    group_by(cluster) %>%
    slice_max(n = 12, order_by = avg_log2FC)

write.csv(top_n, "/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/res/top_genes_per_cluster_new.csv", row.names=FALSE)
write.csv(mo_naive_24hr.markers, "/fh/fast/furlan_s/grp/experiments/timecourse_cd27/MO/res/top_genes_per_cluster_new.csv", row.names=FALSE)

#>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
#Bar plot showing frequency of conditions within each cluster

df<- data.frame(table(mo_naive_24hr$id,mo_naive_24hr$test ))

#df$Var2 <- factor(df$Var2, levels = c("4","5","2","1","3","0")) #changing ordering for bar chart

ggplot(df, aes(fill = Var1, y=Freq, x=Var2)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Clusters") + 
    guides(fill=guide_legend(title="Condition")) +
    scale_fill_manual(values = c("#186272", "red","#5b278c","#bd7ceb","gray"))

df$Var1 <- factor(df$Var1, levels = c("Naive","3-27_24hr","HD_24hr","LD_24hr","BEAD_24hr"))

#Inverting for Carla
ggplot(df, aes(fill = Var2, y=Freq, x=Var1)) + 
    geom_bar(stat = "identity",  position = "fill") +
    theme_classic()+RotatedAxis()+NoGrid() + 
    labs(x = "Condition") + 
    guides(fill=guide_legend(title="Clusters")) +
    scale_fill_manual(values = c("#000000","#17068b","#9915a0","#d7566c","#f79242","#f7e025"))
```

#Top differentially expressed genes

```{r top 12 genes per cluster}
library(svglite)
svglite::svglite("/fh/fast/furlan_s/experiments/timecourse_cd27/MO/figs/cluster_dot_plot.svg", width = 15.5, height = 2)

#DotPlot(mo_naive_24hr, features = unique(top_n$gene), group.by = "test", cluster.idents=F)+scale_y_discrete(limits = c("1","2","3","4","5","6"))+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))&RotatedAxis()

DotPlot(mo_naive_24hr, features = unique(top_n$gene), group.by = "test", cluster.idents=F)+scale_color_gradientn(colors = c("gray70", "#04578f", "#801b1b"))&RotatedAxis()
dev.off()

saveRDS(mo_naive_24hr, file.path(CDS_DIR, "timecourse_MO_cd27_naive_24hr.rds"))
```

#Adding motifs and chromvar assay now that its subset
```{r}

#Motifs------------------------------------------------------------------------------------------
DefaultAssay(mo_naive_24hr)<- "ATAC"

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
mo_naive_24hr <- AddMotifs(
  object = mo_naive_24hr,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm
)
 

human_pwms_v3 <- readRDS("~/cisBP_human_pfms_2021.rds") #Can't find this file?
library(Signac)
DefaultAssay(mo_naive_24hr)<-"ATAC"

mo_naive_24hr@assays$ATAC@motifs<-NULL
mo_naive_24hr <- AddMotifs(
  object = mo_naive_24hr,
  genome =BSgenome.Hsapiens.UCSC.hg38,
  pfm = human_pwms_v3
)

#Chromvar--------------------------------------------------------------------------------

#Runs for a long time
mo_naive_24hr <- RunChromVAR(
  object = mo_naive_24hr,
  genome = BSgenome.Hsapiens.UCSC.hg38
)

rownames(mo_naive_24hr@assays$chromvar)

saveRDS(mo_naive_24hr, file.path(CDS_DIR, "timecourse_MO_cd27_naive_24hr.rds"))
```

#Differential Activity Motifs
```{r DA motifs across exp groups DOT PLOT } 
DefaultAssay(mo_naive_24hr)<-"chromvar"

Idents(mo_naive_24hr)<-"id"

#Runs for a long time
differential.activity <- FindAllMarkers( 
  object = mo_naive_24hr,
  only.pos = T,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)


t<- c()

i=2

for(i in 1:length(differential.activity$gene)){
  t[i]<-mo_naive_24hr@assays$ATAC@motifs@data[[differential.activity$gene[i]]]
}
differential.activity$TF <- t

write.csv(differential.activity, file.path(RES_DIR, "TF_enrichments_group_jaspar.csv"))

differential.activity<-read.csv(file.path(RES_DIR, "TF_enrichments_group_jaspar.csv"))

table(differential.activity$cluster)

#differential.activity<-read.csv(file.path(RES_DIR, "TF_enrichments_group_cisbp.csv"))


#select significant TFS
top_tf<- differential.activity %>%
    dplyr::group_by(cluster) %>%
    slice_max(avg_diff, n = 15)
dim(top_tf)

write.csv(differential.activity, file.path(RES_DIR, "top_TF_enrichments_group_jaspar.csv"))


#get unique pathways
#get values from other groups of top pathways
all_values<- differential.activity[which(differential.activity$TF %in% c(top_tf$TF)),]

all_values$cluster<-factor(all_values$cluster, levels = c("Naive", "24hr_3/27", "24hr_Beads"))
all_values<-all_values %>% mutate(cluster=factor(cluster, levels=c("Naive", "24hr_3/27", "24hr_Beads"))) 
all_values<-all_values %>% arrange(cluster, desc(avg_diff))
all_values$TF<-factor(all_values$TF, levels =rev(unique(all_values$TF)))

##all tfs
#all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(1:27, 43:47, 28:42)])
all_values$TF<-factor(all_values$TF, levels = levels(all_values$TF)[c(10:26, 1:9,27:47)])

ggplot(data = all_values, aes(x = cluster, y = TF)) +
geom_point(aes(size = pct.1, fill = avg_diff), alpha = 0.75, shape = 21) +theme_bw()+NoGrid()+scale_fill_gradientn(colors = viridis(n=100))+RotatedAxis()
```

#Re-embed ATAC and SCT(RNA) UMAPs to see if we can pull apart the experimental groups more
```{r}
#ATAC embedding
DefaultAssay(mo_naive_24hr) <- "ATAC"
mo_naive_24hr <- RunTFIDF(mo_naive_24hr)
mo_naive_24hr <- FindTopFeatures(mo_naive_24hr, min.cutoff = 'q0') #Find most frequently observed features
mo_naive_24hr<- RunSVD(mo_naive_24hr)
ElbowPlot(mo_naive_24hr,reduction = 'lsi')
mo_naive_24hr <- RunUMAP(mo_naive_24hr, reduction = 'lsi', dims = 2:50, reduction.name = "ATAC_UMAP") #running dimensional reduction

DimPlot(mo_naive_24hr,  group.by ="id", label=F, reduction = "ATAC_UMAP",label.box = F,  cols = c("#186272", "#ffb28d", "red","#5b278c","#bd7ceb","gray"))+ggtitle("ATAC UMAP")&NoAxes() #UMAP plot

#SCT embedding
DefaultAssay(mo_naive_24hr) <- "RNA"
mo_naive_24hr<-FindVariableFeatures(mo_naive_24hr) #Identifies features that are outliers on a 'mean variability plot'.
mo_naive_24hr<-ScaleData(mo_naive_24hr) #Scale/Center data
mo_naive_24hr <- SCTransform(mo_naive_24hr, method = "glmGamPoi", vars.to.regress = c("percent.mt","log_RNA")) #A normalization method for single-cell UMI count data using a variance stabilizing transformation.
mo_naive_24hr <- RunPCA(mo_naive_24hr, features = VariableFeatures(mo_naive_24hr))
ElbowPlot(mo_naive_24hr)
ElbowPlot(mo_naive_24hr, ndims = 30)
mo_naive_24hr<-RunUMAP(mo_naive_24hr, dims = 1:50, verbose = T, reduction.name = "SCT_UMAP") #UMAP for SCT method
DimPlot(mo_naive_24hr,  group.by ="id", label=F, reduction = "SCT_UMAP",label.box = F,cols = c("#186272", "#ffb28d", "red","#5b278c","#bd7ceb","gray"))+ggtitle("SCT UMAP")&NoAxes()

#WNN
mo_naive_24hr <- FindMultiModalNeighbors(mo_naive_24hr, reduction.list = list("pca", "lsi"), dims.list = list(1:50, 2:50), weighted.nn.name = "wnn.nn")
mo_naive_24hr <- RunUMAP(mo_naive_24hr, nn.name = "wnn.nn", assay = "ATAC" , reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
DimPlot(mo_naive_24hr, reduction = "wnn.umap")

#umap plots
DimPlot(mo_naive_24hr, group.by = "id", reduction = "wnn.umap", cols = c("#186272", "#ffb28d", "red","#5b278c","#bd7ceb","gray"),label = F, label.box = F)+ggtitle("WNN UMAP")&NoAxes()
DimPlot(mo_naive_24hr, group.by = "dataset", reduction = "wnn.umap", cols = c("#186272", "#ffb28d", "red","#5b278c","#bd7ceb","gray"),label = F, label.box = F)+ggtitle("WNN UMAP")&NoAxes()
DimPlot(mo_naive_24hr, group.by = "timepoint", reduction = "wnn.umap", label = T)+ggtitle("WNN UMAP")
DimPlot(mo_naive_24hr,  group.by ="id", label=T, reduction = "wnn.umap",split.by ="timepoint",label.size = 3,label.box = T)+ggtitle("WNN UMAP")

DefaultAssay(mo_naive_24hr) <- "RNA"
FeaturePlot(mo_naive_24hr, features = "log_ATAC", reduction = "wnn.umap")
VariableFeatures(mo_naive_24hr)



```

#Volcano plots to differentiate 3/28 and 3/27
```{r}
DefaultAssay(mo_naive_24hr) <- "RNA"
rna_markers<-FindMarkers(mo_naive_24hr, group.by = "id", ident.1 = "3-27_24hr", ident.2 = "3-28_24hr",logfc.threshold = 0.1)

#get rid of lnf values
rna_markers$p_val_adj[markers$p_val_adj == 0]<-1e-300

rna_markers$id<-"3/27"
rna_markers$id[rna_markers$avg_log2FC < 0] <- "3/28"

rna_markers$neg_log_pval <- -log10(rna_markers$p_val_adj)

rna_markers$gene<-rownames(rna_markers)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

ggplot(rna_markers, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#186272", "#ffb28d"))+theme_ArchR()+ggtitle("RNA Volcano")

dev.off()
```
#SCT Assay
```{r}
DefaultAssay(mo_naive_24hr) <- "SCT"
SCT_markers<-FindMarkers(mo_naive_24hr, group.by = "id", ident.1 = "3-27_24hr", ident.2 = "3-28_24hr",logfc.threshold = 0.1)

#get rid of lnf values
SCT_markers$p_val_adj[SCT_markers$p_val_adj == 0]<-1e-300

SCT_markers$id<-"3/27"
SCT_markers$id[SCT_markers$avg_log2FC < 0] <- "3/28"

SCT_markers$neg_log_pval <- -log10(SCT_markers$p_val_adj)

SCT_markers$gene<-rownames(SCT_markers)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

ggplot(SCT_markers, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#186272", "#ffb28d"))+theme_ArchR()+ggtitle("SCT Volcano")

dev.off()
```
#GENE_ACC Assay
```{r}
DefaultAssay(mo_naive_24hr) <- "GENE_ACC"
gene_markers<-FindMarkers(mo_naive_24hr, group.by = "id", ident.1 = "3-27_24hr", ident.2 = "3-28_24hr",logfc.threshold = 0.1)

#get rid of lnf values
gene_markers$p_val_adj[gene_markers$p_val_adj == 0]<-1e-300

gene_markers$id<-"3/27"
gene_markers$id[gene_markers$avg_log2FC < 0] <- "3/28"

gene_markers$neg_log_pval <- -log10(gene_markers$p_val_adj)

gene_markers$gene<-rownames(gene_markers)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

ggplot(gene_markers, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#186272", "#ffb28d"))+theme_ArchR()+ggtitle("GENE_ACC Volcano")

dev.off()
```
#ChromvarAssay
```{r}
DefaultAssay(mo_naive_24hr) <- "chromvar"
chromvar_markers<-FindMarkers(mo_naive_24hr, group.by = "id", ident.1 = "3-27_24hr", ident.2 = "3-28_24hr",logfc.threshold = 0.1)

#get rid of lnf values
chromvar_markers$p_val_adj[chromvar_markers$p_val_adj == 0]<-1e-300

chromvar_markers$id<-"3/27"
chromvar_markers$id[chromvar_markers$avg_log2FC < 0] <- "3/28"

chromvar_markers$neg_log_pval <- -log10(chromvar_markers$p_val_adj)

chromvar_markers$gene<-rownames(chromvar_markers)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)

library(ArchR)
library(ggrepel)
options(ggrepel.max.overlaps = 30)
#14x14
ggplot(chromvar_markers, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -0.2 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 0.2 ,linetype= "dashed", color = "gray80")+geom_point(size = 3)+geom_text_repel(color = "black", size =4, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#186272", "#ffb28d"))+theme_ArchR()+ggtitle("ChromVar Volcano")+xlim(-1, 2.5)+ylim(0, 30)+theme(axis.text=element_text(size=12),
        axis.title=element_text(size=14),plot.title = element_text(size = 16, face = "bold"),legend.title=element_text(size=11),legend.text=element_text(size=9))


dev.off()
```


#ChromVar Volcano Plots comparing BEADS vs 3/27 and BEADS vs 3/28
```{r}
DefaultAssay(mo_naive_24hr) <- "chromvar"
bead_vs_27<-FindMarkers(mo_naive_24hr, group.by = "id", ident.1 = "BEAD_24hr", ident.2 = "3-27_24hr",logfc.threshold = 0.1)
bead_vs_28<-FindMarkers(mo_naive_24hr, group.by = "id", ident.1 = "BEAD_24hr", ident.2 = "3-28_24hr",logfc.threshold = 0.1)

#get rid of lnf values
bead_vs_27$p_val_adj[bead_vs_27$p_val_adj == 0]<-1e-300
bead_vs_28$p_val_adj[bead_vs_28$p_val_adj == 0]<-1e-300

bead_vs_27$id <- "BEAD"
bead_vs_27$id[bead_vs_27$avg_log2FC < 0] <- "3/27"

bead_vs_28$id <- "BEAD"
bead_vs_28$id[bead_vs_28$avg_log2FC < 0] <- "3/28"


bead_vs_27$neg_log_pval <- -log10(bead_vs_27$p_val_adj)
bead_vs_28$neg_log_pval <- -log10(bead_vs_28$p_val_adj)


bead_vs_27$gene<-rownames(bead_vs_27)
bead_vs_28$gene<-rownames(bead_vs_28)


options(ggrepel.max.overlaps = 15)

#pdf(file.path(FIG_DIR, "maf_volcano.pdf"), width = 6, height =4)
ggplot(bead_vs_27, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -1.5 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 1.5 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#186272", "red"))+theme_ArchR()+ggtitle("ChromVar Volcano")+xlim(-3, 6)

ggplot(bead_vs_28, aes(x = avg_log2FC, y = neg_log_pval, color = id))+geom_vline(xintercept = 0)+geom_vline(xintercept = -1.5 ,linetype= "dashed", color = "gray80")+geom_vline(xintercept = 1.5 ,linetype= "dashed", color = "gray80")+geom_point()+geom_text_repel(color = "black", size =3, aes(label = gene), min.segment.length = 0)+scale_color_manual(values = c("#ffb28d", "red"))+theme_ArchR()+ggtitle("ChromVar Volcano")+xlim(-3, 6)


dev.off()
```

#Footprinting
```{r}
DefaultAssay(mo_naive_24hr)<-"ATAC"

mo_naive_24hr@assays$ATAC@motifs

unique(sub$Motif) %in% rownames(mo_naive_24hr) 

mo_naive_24hr <- Footprint(
  object = mo_naive_24hr,
  motif.name = unique(sub$Motif),
  genome = BSgenome.Hsapiens.UCSC.hg38
)

DefaultAssay(MO)<-"ATAC"

lapply( 1:length(unique(sub$Motif)), function(x){
  
  pdf(file.path(FIG_DIR, "Footprints",paste0(unique(sub$Motif)[x],"_footprints_PM.pdf")), width =4, height =4) 
  print(PlotFootprint(MO , features = unique(sub$Motif)[x],group.by = "group", normalization = "divide")&scale_color_manual(values = c("Naive" = "gray75", "24hr_3/27" = "#008080", "24hr_Beads" = "Red"))&theme(legend.position = "none"))
  dev.off()
  
})

saveRDS(mo_naive_24hr, file.path(CDS_DIR, "timecourse_MO_cd27_naive_24hr.rds"))
```


# PSEUDOTIME
```{r}
mo_naive_24hr<-readRDS(file.path("/fh/fast/furlan_s/experiments/timecourse_cd27/MO/cds/timecourse_MO_cd27_naive_24hr.rds"))

DefaultAssay(mo_naive_24hr_void_3_28)<-"RNA"

cds<-viewmastR::seurat_to_monocle3(mo_naive_24hr_void_3_28, seu_rd = "wnn.umap")

#troubleshoot reduceDims
cds@int_colData@listData[["reducedDims"]]@listData<-cds@int_colData@listData[["reducedDims"]]@listData[1]

#I usually look at naive genes to pick a "root cell" for trajectories
#the root cell being the least differentiated cell
#i'd set the root node to the bottom edge of cells at the cd27 group of cells


plot_cells(cds, genes = c("TXNIP", "TC2N"), scale_to_range = F) #just chose 2 genes expressed highly in naives
#plot_cells(cds, genes = c("Sell", "Lef1"), scale_to_range = F)

#the cluster resolution will dial the complexity of the trajectory graph
cds<-cluster_cells(cds, resolution = 0.001)
plot_cells(cds)

cds@clusters@listData[["UMAP"]][["clusters"]]<-clusters(cds)
cds<-learn_graph(cds, use_partition = T, close_loop = T)
cds<- order_cells(cds)
plot_cells(cds, color_cells_by = "pseudotime", show_trajectory_graph = F)

mo_naive_24hr_void_3_28$pseudotime<-pseudotime(cds)
pdf("../figs/naive_24hr_void_3_28/pseudotime_umap.pdf", width = 4, height =4)
FeaturePlot(mo_naive_24hr_void_3_28, features = "pseudotime")+scale_color_gradientn(colors = viridis::plasma(n=10))&NoAxes()
dev.off()

pseudo_genes <- c("MKI67","TCF7","SELL","LEF1","IL2RA","MYC","STAT1","TBX21","FOS","JUN","CCR7","PRMT2","CD96","IFNG","TXNIP","FOXO3", "FOXO1", "FOXP1")

pseudo_genes <- c("STAT5B", "BACH1", "BACH2", "MXI1","ATF2","ATF4","FOXJ3","NFATC2","MAZ","NFKB1","NFKB2","ATF7","ATF1","ENO1","GAPDH","ZBTB20","CTCF","EOMES")

rna_genes <- c("TCF7","LEF1","FOXO1","FOXP1","STAT1","CCR7","CD96","ZBTB20","GAPDH","ENO1","IL2RA","IL7R")
                  

pseudo_cds <- cds[rowData(cds)$gene_short_name %in% pseudo_genes,]

pseudo_cds <- cds[rowData(cds)$gene_short_name %in% rna_genes,]

pdf("../figs/genes_pseudotime.pdf", width = 4, height =10)
#9 x 5.00 dimensions
#Gene expression over pseudotime
plot_genes_in_pseudotime(pseudo_cds,
                         color_cells_by="test",
                         min_expr=0.5, cell_size = 1, ncol = 2)+scale_color_manual(values = c("#000000","#17068b","#9915a0","#d7566c","#f79242","#f7e025"))+theme(legend.position = "none")



mo_naive_24hr_void_3_28$pseudotime<-pseudotime(cds)


dev.off()


saveRDS(mo_naive_24hr, file.path(CDS_DIR, "timecourse_MO_cd27_naive_24hr.rds"))
```


#Plotting pseudotime figures
```{r}
library(gridExtra)


mo_naive_24hr$pseudotime<-pseudotime(cds)

seu<-mo_naive_24hr
feature<-"CD96"
assay<-"SCT"
color <-"test"

plot_feature_pseudotime <- function(seu,feature, assay, color) {
  df <- data.frame(pseudotime=seu$pseudotime, seu[[color]])
  feature_vals <- t(seu@assays[[assay]][feature,]) %>% data.frame()
  df$feature_vals = feature_vals[[feature]]
  colnames(df)[2] <- "group_by"
  ggplot(df, aes(x = pseudotime, y = feature_vals, color = group_by))+geom_point(alpha=0.5,size=1)
  }

plot_feature_pseudotime(mo_naive_24hr, "CD96", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.5)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("CD96")+theme(plot.title = element_text(hjust = 0.5))

#10x5
#final colors:
plot_feature_pseudotime(mo_naive_24hr, "CD96", "SCT", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.5)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("CD96")+theme(plot.title = element_text(hjust = 0.5))



#######################################################################################################################################
#RNA Expression
#######################################################################################################################################

ccr7_sct <- plot_feature_pseudotime(mo_naive_24hr, "CCR7", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("CCR7")+theme(plot.title = element_text(hjust = 0.5))

cd96_sct <- plot_feature_pseudotime(mo_naive_24hr, "CD96", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("CD96")+theme(plot.title = element_text(hjust = 0.5))

eomes_sct <- plot_feature_pseudotime(mo_naive_24hr, "EOMES", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("EOMES")+theme(plot.title = element_text(hjust = 0.5))

fos_sct <- plot_feature_pseudotime(mo_naive_24hr, "FOS", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("FOS")+theme(plot.title = element_text(hjust = 0.5))

ifng_sct <- plot_feature_pseudotime(mo_naive_24hr, "IFNG", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("IFNG")+theme(plot.title = element_text(hjust = 0.5))

il2ra_sct <- plot_feature_pseudotime(mo_naive_24hr, "IL2RA", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("IL2RA")+theme(plot.title = element_text(hjust = 0.5))

jun_sct <- plot_feature_pseudotime(mo_naive_24hr, "JUN", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("JUN")+theme(plot.title = element_text(hjust = 0.5))

lef1_sct <- plot_feature_pseudotime(mo_naive_24hr, "LEF1", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("LEF1")+theme(plot.title = element_text(hjust = 0.5))

mki67_sct <- plot_feature_pseudotime(mo_naive_24hr, "MKI67", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("MKI67")+theme(plot.title = element_text(hjust = 0.5))

myc_sct <- plot_feature_pseudotime(mo_naive_24hr, "MYC", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("MYC")+theme(plot.title = element_text(hjust = 0.5))

prmt2_sct <- plot_feature_pseudotime(mo_naive_24hr, "PRMT2", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("PRMT2")+theme(plot.title = element_text(hjust = 0.5))

sell_sct <- plot_feature_pseudotime(mo_naive_24hr, "SELL", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("SELL")+theme(plot.title = element_text(hjust = 0.5))

stat1_sct <- plot_feature_pseudotime(mo_naive_24hr, "STAT1", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("STAT1")+theme(plot.title = element_text(hjust = 0.5))

tbx21_sct <- plot_feature_pseudotime(mo_naive_24hr, "TBX21", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("TBX21")+theme(plot.title = element_text(hjust = 0.5))

tcf7_sct <- plot_feature_pseudotime(mo_naive_24hr, "TCF7", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("TCF7")+theme(plot.title = element_text(hjust = 0.5))

txnip_sct <- plot_feature_pseudotime(mo_naive_24hr, "TXNIP", "SCT", "test")+scale_color_manual(values = c("#A71A1A","#1E96FC", "#16324F","#AC66BA","#C1BEC2", "#EFA00B"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("TXNIP")+theme(plot.title = element_text(hjust = 0.5))

#Dimensions, 7.5 x 6
grid.arrange(ccr7_sct, cd96_sct, eomes_sct, fos_sct, ifng_sct, il2ra_sct, jun_sct, lef1_sct, mki67_sct, myc_sct, prmt2_sct, sell_sct, stat1_sct, tbx21_sct, tcf7_sct, txnip_sct, nrow=8)


#######################################################################################################################################
#ChromVar 
#######################################################################################################################################
ATF1_cv <- plot_feature_pseudotime(mo_naive_24hr, "ATF1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("ATF1")+theme(plot.title = element_text(hjust = 0.5))

ATF2_cv <- plot_feature_pseudotime(mo_naive_24hr, "ATF2", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("ATF2")+theme(plot.title = element_text(hjust = 0.5))

ATF4_cv <- plot_feature_pseudotime(mo_naive_24hr, "ATF4", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("ATF4")+theme(plot.title = element_text(hjust = 0.5))

ATF7_cv <- plot_feature_pseudotime(mo_naive_24hr, "ATF7", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("ATF7")+theme(plot.title = element_text(hjust = 0.5))

BACH1_cv <- plot_feature_pseudotime(mo_naive_24hr, "BACH1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("BACH1")+theme(plot.title = element_text(hjust = 0.5))

BACH2_cv <- plot_feature_pseudotime(mo_naive_24hr, "BACH2", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("BACH2")+theme(plot.title = element_text(hjust = 0.5))

CTCF_cv <- plot_feature_pseudotime(mo_naive_24hr, "CTCF", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("CTCF")+theme(plot.title = element_text(hjust = 0.5))

EOMES_cv <- plot_feature_pseudotime(mo_naive_24hr, "EOMES", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("EOMES")+theme(plot.title = element_text(hjust = 0.5))

FOS_cv <- plot_feature_pseudotime(mo_naive_24hr, "FOS", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("FOS")+theme(plot.title = element_text(hjust = 0.5))

FOXJ3_cv <- plot_feature_pseudotime(mo_naive_24hr, "FOXJ3", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("FOXJ3")+theme(plot.title = element_text(hjust = 0.5))

FOXO1_cv <- plot_feature_pseudotime(mo_naive_24hr, "FOXO1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("FOXO1")+theme(plot.title = element_text(hjust = 0.5))

FOXO3_cv <- plot_feature_pseudotime(mo_naive_24hr, "FOXO3", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("FOXO3")+theme(plot.title = element_text(hjust = 0.5))

FOXP1_cv <- plot_feature_pseudotime(mo_naive_24hr, "FOXP1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("FOXP1")+theme(plot.title = element_text(hjust = 0.5))

JUN_cv <- plot_feature_pseudotime(mo_naive_24hr, "JUN", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("JUN")+theme(plot.title = element_text(hjust = 0.5))

LEF1_cv <- plot_feature_pseudotime(mo_naive_24hr, "LEF1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("LEF1")+theme(plot.title = element_text(hjust = 0.5))

MAZ_cv <- plot_feature_pseudotime(mo_naive_24hr, "MAZ", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("MAZ")+theme(plot.title = element_text(hjust = 0.5))

MXI1_cv <- plot_feature_pseudotime(mo_naive_24hr, "MXI1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("MXI1")+theme(plot.title = element_text(hjust = 0.5))

MYC_cv <- plot_feature_pseudotime(mo_naive_24hr, "MYC", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("MYC")+theme(plot.title = element_text(hjust = 0.5))


NFATC2_cv <- plot_feature_pseudotime(mo_naive_24hr, "NFATC2", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("NFATC2")+theme(plot.title = element_text(hjust = 0.5))

NFKB1_cv <- plot_feature_pseudotime(mo_naive_24hr, "NFKB1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("NFKB1")+theme(plot.title = element_text(hjust = 0.5))

NFKB2_cv <- plot_feature_pseudotime(mo_naive_24hr, "NFKB2", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("NFKB2")+theme(plot.title = element_text(hjust = 0.5))

STAT1_cv <- plot_feature_pseudotime(mo_naive_24hr, "STAT1", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("STAT1")+theme(plot.title = element_text(hjust = 0.5))

STAT5B_cv <- plot_feature_pseudotime(mo_naive_24hr, "STAT5B", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("STAT5B")+theme(plot.title = element_text(hjust = 0.5))

TBX21_cv <- plot_feature_pseudotime(mo_naive_24hr, "TBX21", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("TBX21")+theme(plot.title = element_text(hjust = 0.5))

TCF7_cv <- plot_feature_pseudotime(mo_naive_24hr, "TCF7", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("TCF7")+theme(plot.title = element_text(hjust = 0.5))

ZBTB20_cv <- plot_feature_pseudotime(mo_naive_24hr, "ZBTB20", "chromvar", "test")+scale_color_manual(values = c("#60f7f2","#00A3FF","#090c9b","#00CCFF","#3c3744","#0052FF"))+theme_classic()+ geom_smooth(se = F, color = "Black",linewidth=0.75)+ theme(axis.title = element_blank())+ theme(legend.position = "none")+ggtitle("ZBTB20")+theme(plot.title = element_text(hjust = 0.5))



#Dimensions, 7.5 x 5
grid.arrange(ATF1_cv,ATF2_cv,ATF4_cv,ATF7_cv,BACH1_cv,BACH2_cv,CTCF_cv,EOMES_cv,FOS_cv,FOXJ3_cv,FOXO1_cv,FOXO3_cv,FOXP1_cv,JUN_cv, nrow=7)

#6.5 x 5
grid.arrange(LEF1_cv,MAZ_cv,MXI1_cv,MYC_cv,NFATC2_cv,NFKB1_cv,NFKB2_cv,STAT1_cv,STAT5B_cv,TBX21_cv,TCF7_cv,ZBTB20_cv, nrow=6) 
             


```




